// @salesforce/plugin-templates/lib/commands/lightning/generate/app.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const sf_plugins_core_1 = require("@salesforce/sf-plugins-core");
const lightningAppGenerator_1 = require("@salesforce/templates/lib/generators/lightningAppGenerator");
const utils_1 = require("@salesforce/templates/lib/utils");
const core_1 = require("@salesforce/core");
const templateCommand_1 = require("../../../utils/templateCommand");
const flags_1 = require("../../../utils/flags");
core_1.Messages.importMessagesDirectory(__dirname);
const lightningCommonMessages = core_1.Messages.loadMessages('@salesforce/plugin-templates', 'lightning');
const lightningAppMessages = core_1.Messages.loadMessages('@salesforce/plugin-templates', 'lightningApp');
const lightningAppFileSuffix = /.app$/;
const BUNDLE_TYPE = 'App';
class LightningApp extends sf_plugins_core_1.SfCommand {
    run() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { flags } = yield this.parse(LightningApp);
            const flagsAsOptions = {
                appname: flags.name,
                apiversion: flags['api-version'],
                outputdir: flags['output-dir'],
                template: 'DefaultLightningApp',
                internal: flags.internal,
            };
            return (0, templateCommand_1.runGenerator)({
                generator: lightningAppGenerator_1.default,
                opts: flagsAsOptions,
                ux: new sf_plugins_core_1.Ux({ jsonEnabled: this.jsonEnabled() }),
                templates: (0, templateCommand_1.getCustomTemplates)(this.configAggregator),
            });
        });
    }
}
exports.default = LightningApp;
LightningApp.summary = lightningCommonMessages.getMessage('summary', [BUNDLE_TYPE]);
LightningApp.description = lightningCommonMessages.getMessage('description', [BUNDLE_TYPE]);
LightningApp.examples = lightningAppMessages.getMessages('examples');
LightningApp.aliases = ['force:lightning:app:create'];
LightningApp.deprecateAliases = true;
LightningApp.flags = {
    name: sf_plugins_core_1.Flags.string({
        char: 'n',
        summary: lightningCommonMessages.getMessage('flags.name', [BUNDLE_TYPE]),
        description: lightningCommonMessages.getMessage('flags.name.description'),
        required: true,
        aliases: ['appname'],
        deprecateAliases: true,
    }),
    template: sf_plugins_core_1.Flags.string({
        char: 't',
        summary: lightningCommonMessages.getMessage('flags.template'),
        description: lightningCommonMessages.getMessage('flags.template.description'),
        default: 'DefaultLightningApp',
        options: utils_1.CreateUtil.getCommandTemplatesForFiletype(lightningAppFileSuffix, 'lightningapp'),
    }),
    'output-dir': flags_1.outputDirFlagLightning,
    'api-version': sf_plugins_core_1.orgApiVersionFlagWithDeprecations,
    internal: flags_1.internalFlag,
    loglevel: sf_plugins_core_1.loglevel,
};
