// @salesforce/plugin-templates/lib/commands/lightning/generate/component.js
"use strict";
/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
// tslint:disable:no-unused-expression
const sf_plugins_core_1 = require("@salesforce/sf-plugins-core");
const lightningComponentGenerator_1 = require("@salesforce/templates/lib/generators/lightningComponentGenerator");
const core_1 = require("@salesforce/core");
const templateCommand_1 = require("../../../utils/templateCommand");
const flags_1 = require("../../../utils/flags");
const BUNDLE_TYPE = 'Component';
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-templates', 'lightningCmp');
const lightningCommon = core_1.Messages.loadMessages('@salesforce/plugin-templates', 'lightning');
class LightningComponent extends sf_plugins_core_1.SfCommand {
    run() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { flags } = yield this.parse(LightningComponent);
            const flagsAsOptions = {
                componentname: flags.name,
                template: flags.template,
                outputdir: flags['output-dir'],
                apiversion: flags['api-version'],
                internal: flags.internal,
                type: flags.type,
            };
            return (0, templateCommand_1.runGenerator)({
                generator: lightningComponentGenerator_1.default,
                opts: flagsAsOptions,
                ux: new sf_plugins_core_1.Ux({ jsonEnabled: this.jsonEnabled() }),
                templates: (0, templateCommand_1.getCustomTemplates)(this.configAggregator),
            });
        });
    }
}
exports.default = LightningComponent;
LightningComponent.summary = messages.getMessage('summary');
LightningComponent.description = messages.getMessage('description');
LightningComponent.examples = messages.getMessages('examples');
LightningComponent.aliases = ['force:lightning:component:create'];
LightningComponent.deprecateAliases = true;
LightningComponent.flags = {
    name: sf_plugins_core_1.Flags.string({
        char: 'n',
        summary: lightningCommon.getMessage('flags.name', [BUNDLE_TYPE]),
        description: lightningCommon.getMessage('flags.name.description'),
        required: true,
        aliases: ['componentname'],
        deprecateAliases: true,
    }),
    template: sf_plugins_core_1.Flags.string({
        char: 't',
        summary: lightningCommon.getMessage('flags.template'),
        description: lightningCommon.getMessage('flags.template.description'),
        default: 'default',
        // Note: keep this list here and LightningComponentOptions#template in-sync with the
        // templates/lightningcomponents/[aura|lwc]/* folders
        options: ['default', 'analyticsDashboard', 'analyticsDashboardWithStep'],
    }),
    'output-dir': flags_1.outputDirFlagLightning,
    'api-version': sf_plugins_core_1.orgApiVersionFlagWithDeprecations,
    type: sf_plugins_core_1.Flags.string({
        summary: messages.getMessage('flags.type'),
        options: ['aura', 'lwc'],
        default: 'aura',
    }),
    internal: flags_1.internalFlag,
    loglevel: sf_plugins_core_1.loglevel,
};
