// @salesforce/plugin-templates/lib/commands/lightning/generate/test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const sf_plugins_core_1 = require("@salesforce/sf-plugins-core");
const lightningTestGenerator_1 = require("@salesforce/templates/lib/generators/lightningTestGenerator");
const utils_1 = require("@salesforce/templates/lib/utils");
const core_1 = require("@salesforce/core");
const templateCommand_1 = require("../../../utils/templateCommand");
const flags_1 = require("../../../utils/flags");
const lightningTestFileSuffix = /.resource$/;
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-templates', 'lightningTest');
const lightningMessages = core_1.Messages.loadMessages('@salesforce/plugin-templates', 'lightning');
class LightningTest extends sf_plugins_core_1.SfCommand {
    run() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { flags } = yield this.parse(LightningTest);
            // translate the new flags to the old ones the generator expects
            const flagsAsOptions = {
                testname: flags.name,
                template: 'DefaultLightningTest',
                outputdir: flags['output-dir'],
                internal: flags.internal,
                apiversion: flags['api-version'],
            };
            return (0, templateCommand_1.runGenerator)({
                generator: lightningTestGenerator_1.default,
                opts: flagsAsOptions,
                ux: new sf_plugins_core_1.Ux({ jsonEnabled: this.jsonEnabled() }),
                templates: (0, templateCommand_1.getCustomTemplates)(this.configAggregator),
            });
        });
    }
}
exports.default = LightningTest;
LightningTest.summary = messages.getMessage('summary');
LightningTest.description = messages.getMessage('description');
LightningTest.examples = messages.getMessages('examples');
LightningTest.aliases = ['force:lightning:test:create'];
LightningTest.deprecateAliases = true;
LightningTest.flags = {
    name: sf_plugins_core_1.Flags.string({
        char: 'n',
        summary: lightningMessages.getMessage('flags.name', ['Test']),
        description: messages.getMessage('flags.name.description'),
        required: true,
        aliases: ['testname'],
        deprecateAliases: true,
    }),
    template: sf_plugins_core_1.Flags.string({
        char: 't',
        summary: lightningMessages.getMessage('flags.template'),
        description: lightningMessages.getMessage('flags.template.description'),
        default: 'DefaultLightningTest',
        options: utils_1.CreateUtil.getCommandTemplatesForFiletype(lightningTestFileSuffix, 'lightningtest'),
    }),
    'output-dir': flags_1.outputDirFlagLightning,
    internal: flags_1.internalFlag,
    'api-version': sf_plugins_core_1.orgApiVersionFlagWithDeprecations,
    loglevel: sf_plugins_core_1.loglevel,
};
