// @oclif/plugin-search/dist/commands/search.js
"use strict";
/*
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
const inquirer_1 = require("inquirer");
const core_1 = require("@oclif/core");
const autocomplete_1 = require("../util/autocomplete");
class Search extends core_1.Command {
    async run() {
        const commands = this.config.commands;
        for (const command of commands) {
            command.readableId = (0, core_1.toConfiguredId)(command.id, this.config);
        }
        (0, inquirer_1.registerPrompt)('autocomplete-search', autocomplete_1.AutocompleteSearch);
        const { command } = await (0, inquirer_1.prompt)([
            {
                type: 'autocomplete-search',
                name: 'command',
                message: 'Search for a command',
                emptyText: 'Nothing found!',
                pageSize: 10,
                validate: (val) => {
                    return val ? true : 'Type something!';
                },
                searchOpts: {
                    keys: ['readableId'],
                    shouldSort: true,
                    threshold: 0.25,
                    ignoreLocation: true,
                },
                displayKey: 'readableId',
                options: commands,
                footer: (currentChoices, index) => {
                    const cmd = commands.find(c => c.readableId === currentChoices.getChoice(index).value);
                    if (cmd && cmd.summary)
                        return cmd.summary;
                },
            },
        ]);
        if (command === 'help') {
            return this.config.runCommand((0, core_1.toStandardizedId)(command, this.config), []);
        }
        return this.config.runCommand('help', [(0, core_1.toStandardizedId)(command, this.config)]);
    }
}
exports.default = Search;
Search.summary = 'Search for a command.';
Search.description = 'Once you select a command, hit enter and it will show the help for that command.';
