// @salesforce/plugin-packaging/lib/commands/force/package/beta/version/delete.js
"use strict";
/*
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PackageVersionDeleteCommand = void 0;
const os = require("os");
const command_1 = require("@salesforce/command");
const core_1 = require("@salesforce/core");
const packaging_1 = require("@salesforce/packaging");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-packaging', 'package_version_delete');
class PackageVersionDeleteCommand extends command_1.SfdxCommand {
    async run() {
        const packageVersion = new packaging_1.PackageVersion({
            connection: this.hubOrg.getConnection(),
            project: this.project,
            idOrAlias: this.flags.package,
        });
        await this.confirmDelete();
        const results = this.flags.undelete ? await packageVersion.undelete() : await packageVersion.delete();
        this.ux.log(this.getHumanSuccessMessage(results));
        return results;
    }
    async confirmDelete() {
        if (this.flags.noprompt || this.flags.json) {
            return true;
        }
        const message = this.flags.undelete ? messages.getMessage('promptUndelete') : messages.getMessage('promptDelete');
        const accepted = await this.ux.confirm(message);
        if (!accepted) {
            throw new Error(messages.getMessage('promptDeleteDeny'));
        }
    }
    getHumanSuccessMessage(result) {
        return messages.getMessage(this.flags.undelete ? 'humanSuccessUndelete' : 'humanSuccess', [result.id]);
    }
}
exports.PackageVersionDeleteCommand = PackageVersionDeleteCommand;
PackageVersionDeleteCommand.description = messages.getMessage('cliDescription');
PackageVersionDeleteCommand.examples = messages.getMessage('examples').split(os.EOL);
PackageVersionDeleteCommand.requiresDevhubUsername = true;
PackageVersionDeleteCommand.requiresProject = true;
PackageVersionDeleteCommand.flagsConfig = {
    noprompt: command_1.flags.boolean({
        char: 'n',
        description: messages.getMessage('noPrompt'),
        longDescription: messages.getMessage('noPrompt'),
    }),
    package: command_1.flags.string({
        char: 'p',
        description: messages.getMessage('package'),
        longDescription: messages.getMessage('packageLong'),
        required: true,
    }),
    undelete: command_1.flags.boolean({
        description: messages.getMessage('undelete'),
        longDescription: messages.getMessage('undeleteLong'),
        hidden: true,
    }),
};
