// @salesforce/plugin-packaging/lib/commands/force/package/beta/version/displayancestry.js
"use strict";
/*
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PackageVersionDisplayAncestryCommand = void 0;
const os = require("os");
const command_1 = require("@salesforce/command");
const core_1 = require("@salesforce/core");
const packaging_1 = require("@salesforce/packaging");
// Import i18n messages
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-packaging', 'package_displayancestry');
class PackageVersionDisplayAncestryCommand extends command_1.SfdxCommand {
    async run() {
        const packageAncestry = await packaging_1.Package.getAncestry(this.flags.package, this.project, this.hubOrg.getConnection());
        const jsonProducer = packageAncestry.getJsonProducer();
        if (this.flags.dotcode) {
            const dotProducer = packageAncestry.getDotProducer();
            const dotCodeResult = dotProducer.produce();
            if (this.flags.json) {
                return dotCodeResult;
            }
            else {
                this.ux.log(dotCodeResult);
            }
        }
        else {
            if (packageAncestry.requestedPackageId.startsWith('04t')) {
                const paths = packageAncestry.getLeafPathToRoot(packageAncestry.requestedPackageId);
                this.ux.log(`${paths[0].map((p) => p.getVersion()).join(' -> ')} (root)`);
                this.ux.log();
            }
            const treeProducer = packageAncestry.getTreeProducer(!!this.flags.verbose);
            if (!this.flags.json) {
                treeProducer.produce();
            }
        }
        return jsonProducer.produce();
    }
}
exports.PackageVersionDisplayAncestryCommand = PackageVersionDisplayAncestryCommand;
PackageVersionDisplayAncestryCommand.description = messages.getMessage('cliDescription');
PackageVersionDisplayAncestryCommand.examples = messages.getMessage('examples').split(os.EOL);
PackageVersionDisplayAncestryCommand.showProgress = false;
PackageVersionDisplayAncestryCommand.varargs = false;
PackageVersionDisplayAncestryCommand.requiresDevhubUsername = true;
PackageVersionDisplayAncestryCommand.requiresProject = true;
PackageVersionDisplayAncestryCommand.flagsConfig = {
    // --json is configured automatically
    package: command_1.flags.string({
        char: 'p',
        description: messages.getMessage('package'),
        longDescription: messages.getMessage('packageLong'),
        required: true,
    }),
    dotcode: command_1.flags.boolean({
        description: messages.getMessage('dotcode'),
        longDescription: messages.getMessage('dotcodeLong'),
    }),
    verbose: command_1.flags.builtin({
        description: messages.getMessage('verbose'),
        longDescription: messages.getMessage('verboseLong'),
    }),
};
