// @salesforce/plugin-org/lib/commands/org/display.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrgDisplayCommand = void 0;
const sf_plugins_core_1 = require("@salesforce/sf-plugins-core");
const core_1 = require("@salesforce/core");
const kit_1 = require("@salesforce/kit");
const utils_1 = require("../../shared/utils");
const orgHighlighter_1 = require("../../shared/orgHighlighter");
const orgListUtil_1 = require("../../shared/orgListUtil");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-org', 'display');
const sharedMessages = core_1.Messages.loadMessages('@salesforce/plugin-org', 'messages');
class OrgDisplayCommand extends sf_plugins_core_1.SfCommand {
    async run() {
        const { flags } = await this.parse(OrgDisplayCommand);
        this.org = flags['target-org'];
        this.org.getConnection(flags['api-version']);
        try {
            // the auth file might have a stale access token.  We want to refresh it before getting the fields
            await this.org.refreshAuth();
        }
        catch (error) {
            // even if this fails, we want to display the information we can read from the auth file
            this.warn('unable to refresh auth for org');
        }
        // translate to alias if necessary
        const authInfo = await core_1.AuthInfo.create({ username: this.org.getUsername() });
        const fields = authInfo.getFields(true);
        const isScratchOrg = Boolean(fields.devHubUsername);
        const scratchOrgInfo = isScratchOrg && fields.orgId ? await this.getScratchOrgInformation(fields.orgId) : {};
        const returnValue = {
            // renamed properties
            id: fields.orgId,
            devHubId: fields.devHubUsername,
            // copied properties
            apiVersion: fields.instanceApiVersion,
            accessToken: fields.accessToken,
            instanceUrl: fields.instanceUrl,
            username: fields.username,
            clientId: fields.clientId,
            password: fields.password,
            ...scratchOrgInfo,
            // properties with more complex logic
            connectedStatus: isScratchOrg
                ? undefined
                : await orgListUtil_1.OrgListUtil.determineConnectedStatusForNonScratchOrg(fields.username),
            sfdxAuthUrl: flags.verbose && fields.refreshToken ? authInfo.getSfdxAuthUrl() : undefined,
            alias: await (0, utils_1.getAliasByUsername)(fields.username),
        };
        this.warn(sharedMessages.getMessage('SecurityWarning'));
        this.print(returnValue);
        return returnValue;
    }
    print(result) {
        this.log();
        const tableRows = Object.entries(result)
            .filter(([, value]) => value !== undefined && value !== null) // some values won't exist
            .sort() // this command always alphabetizes the table rows
            .map(([key, value]) => ({
            key: (0, kit_1.camelCaseToTitleCase)(key),
            value: typeof value === 'string' ? (0, orgHighlighter_1.getStyledValue)(key, value) : value,
        }));
        this.styledHeader('Org Description');
        this.table(tableRows, {
            key: { header: 'KEY' },
            value: { header: 'VALUE' },
        });
    }
    async getScratchOrgInformation(orgId) {
        const hubOrg = await this.org.getDevHubOrg();
        // we know this is a scratch org so it must have a hubOrg and that'll have a username
        const hubUsername = hubOrg?.getUsername();
        const result = (await orgListUtil_1.OrgListUtil.retrieveScratchOrgInfoFromDevHub(hubUsername, [(0, core_1.trimTo15)(orgId)]))[0];
        if (result) {
            return {
                status: result.Status,
                devHubId: hubUsername,
                expirationDate: result.ExpirationDate,
                createdBy: result.CreatedBy?.Username,
                edition: result.Edition ?? undefined,
                namespace: result.Namespace ?? undefined,
                orgName: result.OrgName,
                createdDate: result.CreatedDate,
                signupUsername: result.SignupUsername,
            };
        }
        throw new core_1.SfError(messages.getMessage('noScratchOrgInfoError', [core_1.sfdc.trimTo15(orgId), hubUsername]), 'NoScratchInfo', [messages.getMessage('noScratchOrgInfoAction')]);
    }
}
exports.OrgDisplayCommand = OrgDisplayCommand;
OrgDisplayCommand.summary = messages.getMessage('summary');
OrgDisplayCommand.description = messages.getMessage('description');
OrgDisplayCommand.examples = messages.getMessages('examples');
OrgDisplayCommand.aliases = ['force:org:display'];
OrgDisplayCommand.deprecateAliases = true;
OrgDisplayCommand.flags = {
    'target-org': sf_plugins_core_1.requiredOrgFlagWithDeprecations,
    'api-version': sf_plugins_core_1.orgApiVersionFlagWithDeprecations,
    verbose: sf_plugins_core_1.Flags.boolean({
        summary: messages.getMessage('flags.verbose.summary'),
    }),
    loglevel: sf_plugins_core_1.loglevel,
};
