// @salesforce/plugin-templates/lib/commands/visualforce/generate/component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const sf_plugins_core_1 = require("@salesforce/sf-plugins-core");
const visualforceComponentGenerator_1 = require("@salesforce/templates/lib/generators/visualforceComponentGenerator");
const utils_1 = require("@salesforce/templates/lib/utils");
const core_1 = require("@salesforce/core");
const flags_1 = require("../../../utils/flags");
const templateCommand_1 = require("../../../utils/templateCommand");
const visualforceComponentFileSuffix = /.component$/;
const VF_TYPE = 'Component';
core_1.Messages.importMessagesDirectory(__dirname);
const commonMessages = core_1.Messages.loadMessages('@salesforce/plugin-templates', 'messages');
const messages = core_1.Messages.loadMessages('@salesforce/plugin-templates', 'vf');
class VisualforceComponent extends sf_plugins_core_1.SfCommand {
    run() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { flags } = yield this.parse(VisualforceComponent);
            // translate the new flags to the old ones the generator expects
            const flagsAsOptions = {
                componentname: flags.name,
                label: flags.label,
                template: 'DefaultVFComponent',
                outputdir: flags['output-dir'],
                apiversion: flags['api-version'],
            };
            return (0, templateCommand_1.runGenerator)({
                generator: visualforceComponentGenerator_1.default,
                opts: flagsAsOptions,
                ux: new sf_plugins_core_1.Ux({ jsonEnabled: this.jsonEnabled() }),
                templates: (0, templateCommand_1.getCustomTemplates)(this.configAggregator),
            });
        });
    }
}
exports.default = VisualforceComponent;
VisualforceComponent.summary = messages.getMessage('summary', [VF_TYPE]);
VisualforceComponent.description = messages.getMessage('description', [VF_TYPE]);
VisualforceComponent.examples = messages.getMessages('examples.component');
VisualforceComponent.aliases = ['force:visualforce:component:create'];
VisualforceComponent.deprecateAliases = true;
VisualforceComponent.flags = {
    name: sf_plugins_core_1.Flags.string({
        char: 'n',
        summary: messages.getMessage('flags.name', [VF_TYPE]),
        description: messages.getMessage('flags.name.description'),
        required: true,
        aliases: ['componentname'],
        deprecateAliases: true,
    }),
    template: sf_plugins_core_1.Flags.string({
        char: 't',
        summary: commonMessages.getMessage('flags.template'),
        description: commonMessages.getMessage('flags.template.description'),
        default: 'DefaultVFComponent',
        options: utils_1.CreateUtil.getCommandTemplatesForFiletype(visualforceComponentFileSuffix, 'visualforcecomponent'),
    }),
    'output-dir': flags_1.outputDirFlag,
    'api-version': sf_plugins_core_1.orgApiVersionFlagWithDeprecations,
    label: sf_plugins_core_1.Flags.string({
        char: 'l',
        summary: messages.getMessage('flags.label', [VF_TYPE]),
        required: true,
    }),
    loglevel: sf_plugins_core_1.loglevel,
};
