// @salesforce/plugin-deploy-retrieve/lib/formatters/metadataConvertResultFormatter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetadataConvertResultFormatter = void 0;
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const path = require("path");
const core_1 = require("@oclif/core");
class MetadataConvertResultFormatter {
    constructor(result) {
        this.result = result;
    }
    getJson() {
        this.convertResults = [];
        this.result?.converted?.forEach((component) => {
            if (component.xml) {
                this.convertResults.push({
                    fullName: component.fullName,
                    type: component.type.name,
                    filePath: path.relative('.', component.xml),
                    state: 'Add',
                });
            }
            if (component.content) {
                this.convertResults.push({
                    fullName: component.fullName,
                    type: component.type.name,
                    filePath: path.relative('.', component.content),
                    state: 'Add',
                });
            }
        });
        return this.convertResults;
    }
    display() {
        const convertData = this.getJson();
        if (convertData?.length) {
            core_1.ux.table(convertData.map((entry) => ({
                state: entry.state,
                fullName: entry.fullName,
                type: entry.type,
                filePath: entry.filePath,
            })), {
                state: { header: 'STATE' },
                fullName: { header: 'FULL NAME' },
                type: { header: 'TYPE' },
                filePath: { header: 'PROJECT PATH' },
            });
        }
        else {
            core_1.ux.log('No metadata found to convert');
        }
    }
}
exports.MetadataConvertResultFormatter = MetadataConvertResultFormatter;
