// @salesforce/plugin-deploy-retrieve/lib/commands/project/delete/tracking.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeleteTracking = void 0;
const core_1 = require("@salesforce/core");
const chalk = require("chalk");
const source_tracking_1 = require("@salesforce/source-tracking");
const sf_plugins_core_1 = require("@salesforce/sf-plugins-core");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-deploy-retrieve', 'delete.tracking');
class DeleteTracking extends sf_plugins_core_1.SfCommand {
    async run() {
        const { flags } = await this.parse(DeleteTracking);
        let clearedFiles = [];
        if (flags['no-prompt'] || (await this.confirm(chalk.dim(messages.getMessage('promptMessage'))))) {
            const sourceTracking = await source_tracking_1.SourceTracking.create({
                project: this.project,
                org: flags['target-org'],
            });
            clearedFiles = await Promise.all([sourceTracking.clearLocalTracking(), sourceTracking.clearRemoteTracking()]);
            this.logSuccess('Cleared local tracking files.');
        }
        return { clearedFiles };
    }
}
exports.DeleteTracking = DeleteTracking;
DeleteTracking.deprecateAliases = true;
DeleteTracking.aliases = ['force:source:tracking:clear'];
DeleteTracking.summary = messages.getMessage('deleteSummary');
DeleteTracking.description = messages.getMessage('deleteDescription');
DeleteTracking.requiresProject = true;
DeleteTracking.examples = [];
DeleteTracking.flags = {
    'api-version': sf_plugins_core_1.orgApiVersionFlagWithDeprecations,
    loglevel: sf_plugins_core_1.loglevel,
    'target-org': sf_plugins_core_1.requiredOrgFlagWithDeprecations,
    'no-prompt': sf_plugins_core_1.Flags.boolean({
        char: 'p',
        summary: messages.getMessage('flags.no-prompt.summary'),
        aliases: ['noprompt'],
        deprecateAliases: true,
    }),
};
