// @salesforce/plugin-user/lib/commands/org/generate/password.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenerateUserPasswordCommand = void 0;
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const sf_plugins_core_1 = require("@salesforce/sf-plugins-core");
const core_1 = require("@salesforce/core");
const kit_1 = require("@salesforce/kit");
const generate_1 = require("../../../baseCommands/user/password/generate");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-user', 'password.generate');
class GenerateUserPasswordCommand extends generate_1.UserPasswordGenerateBaseCommand {
    async run() {
        const { flags } = await this.parse(GenerateUserPasswordCommand);
        this.usernames = (0, kit_1.ensureArray)(flags['on-behalf-of'] ?? flags['target-org'].getUsername());
        this.length = flags.length;
        this.complexity = flags.complexity;
        this.org = flags['target-org'];
        this.connection = this.org.getConnection(flags['api-version']);
        return this.generate();
    }
}
exports.GenerateUserPasswordCommand = GenerateUserPasswordCommand;
GenerateUserPasswordCommand.summary = messages.getMessage('summary');
GenerateUserPasswordCommand.description = messages.getMessage('description');
GenerateUserPasswordCommand.examples = messages.getMessages('examples');
GenerateUserPasswordCommand.flags = {
    'on-behalf-of': sf_plugins_core_1.Flags.string({
        aliases: ['onbehalfof'],
        deprecateAliases: true,
        char: 'b',
        summary: messages.getMessage('flags.onBehalfOf.summary'),
        multiple: true,
        parse: (input) => {
            if (input.includes(',')) {
                throw messages.createError('onBehalfOfMultipleError');
            }
            return Promise.resolve(input);
        },
    }),
    length: sf_plugins_core_1.Flags.integer({
        char: 'l',
        summary: messages.getMessage('flags.length.summary'),
        min: 8,
        max: 1000,
        default: 13,
    }),
    // the higher the value, the stronger the password
    complexity: sf_plugins_core_1.Flags.integer({
        char: 'c',
        summary: messages.getMessage('flags.complexity.summary'),
        min: 0,
        max: 5,
        default: 5,
    }),
    'target-org': sf_plugins_core_1.Flags.requiredOrg({ required: true }),
    'api-version': sf_plugins_core_1.Flags.orgApiVersion(),
};
