// @salesforce/plugin-signups/lib/commands/org/delete/snapshot.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SnapshotDelete = void 0;
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const os_1 = require("os");
const sf_plugins_core_1 = require("@salesforce/sf-plugins-core");
const core_1 = require("@salesforce/core");
const snapshot_1 = require("../../../shared/snapshot");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-signups', 'snapshot.delete');
// jsforce can return SaveError[] or never[]
const isSaveError = (error) => error.message !== undefined;
class SnapshotDelete extends sf_plugins_core_1.SfCommand {
    async run() {
        // resolve the query to an ID.  This also verifies the snapshot exists in the org
        const { flags } = await this.parse(SnapshotDelete);
        const conn = flags['target-dev-hub'].getConnection(flags['api-version']);
        const result = await (0, snapshot_1.queryByNameOrId)(conn, flags.snapshot);
        const deleteResult = await conn.sobject('OrgSnapshot').delete(result.Id);
        if (deleteResult.success) {
            this.logSuccess(messages.getMessage('success', [flags.snapshot]));
            return deleteResult;
        }
        throw new Error(deleteResult.errors
            .filter(isSaveError)
            .map((error) => error.message)
            .join(os_1.EOL));
    }
}
exports.SnapshotDelete = SnapshotDelete;
SnapshotDelete.summary = messages.getMessage('summary');
SnapshotDelete.description = messages.getMessage('description');
SnapshotDelete.examples = messages.getMessages('examples');
SnapshotDelete.aliases = ['force:org:snapshot:delete'];
SnapshotDelete.deprecateAliases = true;
SnapshotDelete.flags = {
    'target-dev-hub': sf_plugins_core_1.requiredHubFlagWithDeprecations,
    'api-version': sf_plugins_core_1.orgApiVersionFlagWithDeprecations,
    loglevel: sf_plugins_core_1.loglevel,
    snapshot: sf_plugins_core_1.Flags.string({
        char: 's',
        summary: messages.getMessage('flags.snapshot.summary'),
        description: messages.getMessage('flags.snapshot.description'),
        required: true,
    }),
};
