// @salesforce/plugin-settings/lib/commands/config/get.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Get = void 0;
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const sf_plugins_core_1 = require("@salesforce/sf-plugins-core");
const core_1 = require("@salesforce/core");
const config_1 = require("../../config");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-settings', 'config.get');
class Get extends config_1.ConfigCommand {
    async run() {
        const { argv, flags } = await this.parse(Get);
        // instantiate a SfdxConfigAggregator to get the restDeploy <-> org-metadata-rest-deploy deprecation linked
        await core_1.SfdxConfigAggregator.create({});
        if (!argv || argv.length === 0) {
            throw messages.createError('error.NoConfigKeysFound');
        }
        const aggregator = await core_1.ConfigAggregator.create();
        for (const configName of argv) {
            try {
                this.pushSuccess(aggregator.getInfo(configName, true));
            }
            catch (err) {
                const error = err;
                if (error.message.includes('Deprecated config name')) {
                    // because we've caught the deprecated error, the 'newKey' property will be set
                    const info = aggregator.getInfo(aggregator.getPropertyMeta(configName).newKey);
                    // deprecated key, so we'll get the replacement and return the value
                    this.responses.push({
                        name: info.key,
                        key: configName,
                        value: info.value,
                        deprecated: true,
                        location: info.location,
                        path: info.path,
                        error,
                        message: error.message,
                        success: true,
                    });
                }
                else if (error.name.includes('UnknownConfigKeyError') && !this.jsonEnabled()) {
                    const suggestion = this.calculateSuggestion(configName);
                    // eslint-disable-next-line no-await-in-loop
                    const answer = (await this.confirm(messages.getMessage('didYouMean', [suggestion]), 10 * 1000)) ?? false;
                    if (answer) {
                        this.pushSuccess(aggregator.getInfo(suggestion, false));
                    }
                }
                else {
                    this.pushFailure(configName, err);
                }
            }
        }
        this.output('Get Config', flags.verbose);
        return this.responses;
    }
}
exports.Get = Get;
Get.description = messages.getMessage('description');
Get.summary = messages.getMessage('summary');
Get.examples = messages.getMessages('examples');
Get.aliases = ['force:config:get'];
Get.deprecateAliases = true;
Get.strict = false;
Get.flags = {
    loglevel: sf_plugins_core_1.loglevel,
    verbose: sf_plugins_core_1.Flags.boolean({
        summary: messages.getMessage('flags.verbose.summary'),
    }),
};
Get.configurationVariablesSection = config_1.CONFIG_HELP_SECTION;
