// @salesforce/plugin-data/lib/BulkBaseCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BulkBaseCommand = void 0;
/*
 * Copyright (c) 2023, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const sf_plugins_core_1 = require("@salesforce/sf-plugins-core");
const kit_1 = require("@salesforce/kit");
const change_case_1 = require("change-case");
const core_1 = require("@salesforce/core");
const reporters_1 = require("./reporters");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-data', 'bulk.base.command');
class BulkBaseCommand extends sf_plugins_core_1.SfCommand {
    constructor() {
        super(...arguments);
        this.lifeCycle = core_1.Lifecycle.getInstance();
        this.isAsync = false;
        this.endWaitTime = 0;
        this.wait = 0;
        this.numberRecordsProcessed = 0;
        this.numberRecordsFailed = 0;
        this.numberRecordSuceeded = 0;
        this.timeout = false;
    }
    displayBulkV2Result(jobInfo) {
        if (this.isAsync) {
            this.logSuccess(messages.getMessage('success', [this.operation, jobInfo.id]));
            this.info(messages.getMessage('checkStatus', [
                this.config.bin,
                this.operation,
                jobInfo.id,
                this.connection?.getUsername(),
            ]));
        }
        else {
            this.log();
            this.info((0, reporters_1.getResultMessage)(jobInfo));
            if ((jobInfo.numberRecordsFailed ?? 0) > 0 || jobInfo.state === 'Failed') {
                this.info(messages.getMessage('checkJobViaUi', [this.config.bin, this.connection?.getUsername(), jobInfo.id]));
            }
            if (jobInfo.state === 'InProgress' || jobInfo.state === 'Open') {
                this.info(messages.getMessage('checkStatus', [
                    this.config.bin,
                    this.operation,
                    jobInfo.id,
                    this.connection?.getUsername(),
                ]));
            }
            if (jobInfo.state === 'Failed') {
                throw messages.createError('bulkJobFailed', [jobInfo.id]);
            }
        }
    }
    setupLifecycleListeners() {
        // eslint-disable-next-line @typescript-eslint/no-misused-promises
        this.job.on('jobProgress', async () => {
            const jobInfo = await this.job.check();
            this.numberRecordsProcessed = jobInfo.numberRecordsProcessed ?? 0;
            this.numberRecordsFailed = jobInfo.numberRecordsFailed ?? 0;
            this.numberRecordSuceeded = this.numberRecordsProcessed - this.numberRecordsFailed;
            this.spinner.status = `${this.getRemainingTimeStatus()}${this.getStage(jobInfo.state)}${this.getRemainingRecordsStatus()}`;
        });
        this.job.on('error', (message) => {
            try {
                this.error(message);
            }
            finally {
                this.spinner.stop();
            }
        });
        // eslint-disable-next-line @typescript-eslint/no-misused-promises
        this.job.on('jobTimeout', async () => {
            if (!this.timeout) {
                this.timeout = true;
                await this.cache?.createCacheEntryForRequest(this.job.id ?? '', this.connection?.getUsername(), this.connection?.getApiVersion());
                this.displayBulkV2Result(await this.job.check());
            }
        });
    }
    getRemainingTimeStatus() {
        return this.isAsync
            ? ''
            : messages.getMessage('remainingTimeStatus', [kit_1.Duration.milliseconds(this.endWaitTime - Date.now()).minutes]);
    }
    getRemainingRecordsStatus() {
        // the leading space is intentional
        return ` ${messages.getMessage('remainingRecordsStatus', [
            this.numberRecordSuceeded,
            this.numberRecordsFailed,
            this.numberRecordsProcessed,
        ])}`;
    }
    // eslint-disable-next-line class-methods-use-this
    getStage(state) {
        return ` Stage: ${(0, change_case_1.capitalCase)(state)}.`;
    }
}
exports.BulkBaseCommand = BulkBaseCommand;
