// @salesforce/plugin-source/lib/commands/force/source/status.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@salesforce/core");
const sf_plugins_core_1 = require("@salesforce/sf-plugins-core");
const statusFormatter_1 = require("../../../formatters/source/statusFormatter");
const trackingFunctions_1 = require("../../../trackingFunctions");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-source', 'status');
const replacement = 'project retrieve/deploy preview';
class Status extends sf_plugins_core_1.SfCommand {
    constructor() {
        super(...arguments);
        this.results = new Array();
        this.localAdds = [];
    }
    async run() {
        this.flags = (await this.parse(Status)).flags;
        const tracking = await (0, trackingFunctions_1.trackingSetup)({
            ignoreConflicts: true,
            org: this.flags['target-org'],
            project: this.project,
            ux: new sf_plugins_core_1.Ux({ jsonEnabled: this.jsonEnabled() }),
        });
        const wantsLocal = this.flags.local || (!this.flags.remote && !this.flags.local);
        const wantsRemote = this.flags.remote || (!this.flags.remote && !this.flags.local);
        this.debug(`project is ${this.project.getPath()} and pkgDirs are ${this.project
            .getPackageDirectories()
            .map((dir) => dir.path)
            .join(',')}`);
        const stlStatusResult = await tracking.getStatus({ local: wantsLocal, remote: wantsRemote });
        this.results = stlStatusResult.map((result) => resultConverter(result));
        return this.formatResult();
    }
    formatResult() {
        const formatter = new statusFormatter_1.StatusFormatter(new sf_plugins_core_1.Ux({ jsonEnabled: this.jsonEnabled() }), { concise: this.flags.concise }, this.results);
        if (!this.flags.json) {
            formatter.display();
        }
        return formatter.getJson();
    }
}
exports.default = Status;
Status.summary = messages.getMessage('summary');
Status.examples = messages.getMessages('examples');
Status.state = 'deprecated';
Status.deprecationOptions = {
    to: replacement,
    message: messages.getMessage('deprecation', [replacement]),
};
Status.flags = {
    'api-version': sf_plugins_core_1.orgApiVersionFlagWithDeprecations,
    loglevel: sf_plugins_core_1.loglevel,
    'target-org': sf_plugins_core_1.requiredOrgFlagWithDeprecations,
    local: sf_plugins_core_1.Flags.boolean({
        char: 'l',
        summary: messages.getMessage('flags.local.summary'),
        exclusive: ['remote'],
    }),
    remote: sf_plugins_core_1.Flags.boolean({
        char: 'r',
        summary: messages.getMessage('flags.remote.summary'),
        exclusive: ['local'],
    }),
    concise: sf_plugins_core_1.Flags.boolean({
        summary: messages.getMessage('flags.concise.summary'),
    }),
};
Status.requiresProject = true;
/**
 * STL provides a more useful json output.
 * This function makes it consistent with the Status command's json.
 */
const resultConverter = (input) => {
    const { fullName, type, ignored, filePath, conflict } = input;
    const origin = originMap.get(input.origin);
    const actualState = stateMap.get(input.state);
    return {
        fullName,
        type,
        // this string became the place to store information.
        // The JSON now breaks out that info but preserves this property for backward compatibility
        state: `${origin} ${actualState}${conflict ? ' (Conflict)' : ''}`,
        ignored,
        filePath,
        origin,
        actualState,
        conflict,
    };
};
const originMap = new Map([
    ['local', 'Local'],
    ['remote', 'Remote'],
]);
const stateMap = new Map([
    ['delete', 'Deleted'],
    ['add', 'Add'],
    ['modify', 'Changed'],
    ['nondelete', 'Changed'],
]);
