// @salesforce/plugin-trust/lib/hooks/jitPluginInstall.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const os = require("os");
const core_1 = require("@salesforce/core");
const appInsights_1 = require("@salesforce/telemetry/lib/appInsights");
const hook = async function (opts) {
    try {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
        global.cliTelemetry?.record({
            eventName: 'JIT_INSTALL_STARTED',
            type: 'EVENT',
            version: opts.pluginVersion,
            plugin: opts.command.pluginName,
            command: opts.command.id,
        });
        await opts.config.runCommand('plugins:install', [`${opts.command.pluginName}@${opts.pluginVersion}`]);
        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
        global.cliTelemetry?.record({
            eventName: 'JIT_INSTALL_SUCCESS',
            type: 'EVENT',
            version: opts.pluginVersion,
            plugin: opts.command.pluginName,
            command: opts.command.id,
        });
    }
    catch (error) {
        const err = error;
        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
        global.cliTelemetry?.record({
            eventName: 'JIT_INSTALL_FAILED',
            type: 'EVENT',
            message: err.message,
            // eslint-disable-next-line @typescript-eslint/no-unsafe-argument, @typescript-eslint/no-unsafe-member-access
            stackTrace: err?.stack?.replace(new RegExp(os.homedir(), 'g'), appInsights_1.AppInsights.GDPR_HIDDEN),
            version: opts.pluginVersion,
            plugin: opts.command.pluginName,
            command: opts.command.id,
        });
        throw new core_1.SfError(`Could not install ${opts.command.pluginName}`, 'JitPluginInstallError');
    }
};
exports.default = hook;
