// @salesforce/packaging/lib/interfaces/packagingInterfacesAndType.js
"use strict";
/*
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Package1VersionEvents = exports.PackageVersionEvents = exports.PackageEvents = exports.PackageVersionCreateRequestResultInProgressStatuses = void 0;
const packagingSObjects_1 = require("./packagingSObjects");
var Package2VersionStatus = packagingSObjects_1.PackagingSObjects.Package2VersionStatus;
exports.PackageVersionCreateRequestResultInProgressStatuses = Object.values(Package2VersionStatus).filter((status) => !['Queued', 'Success', 'Error'].includes(status));
exports.PackageEvents = {
    convert: {
        success: 'Package/convert-success',
        error: 'Package/convert-error',
        progress: 'Package/convert-in-progress',
    },
    install: {
        warning: 'Package/install-warning',
        presend: 'Package/install-presend',
        postsend: 'Package/install-postsend',
        status: 'Package/install-status',
        'subscriber-status': 'Package/install-subscriber-status',
    },
    uninstall: 'Package/uninstall',
};
exports.PackageVersionEvents = {
    create: {
        enqueued: 'PackageVersion/create-enqueued',
        progress: 'PackageVersion/create-in-progress',
        success: 'PackageVersion/create-success',
        error: 'PackageVersion/create-error',
        'timed-out': 'PackageVersion/create-timed-out',
        'preserve-files': 'PackageVersion/create-preserve-files',
    },
};
exports.Package1VersionEvents = {
    create: {
        progress: 'Package1Version/create-progress',
    },
};
