// @salesforce/packaging/lib/interfaces/packagingSObjects.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PackagingSObjects = void 0;
var PackagingSObjects;
(function (PackagingSObjects) {
    let Package2VersionStatus;
    (function (Package2VersionStatus) {
        Package2VersionStatus["queued"] = "Queued";
        Package2VersionStatus["inProgress"] = "InProgress";
        Package2VersionStatus["success"] = "Success";
        Package2VersionStatus["error"] = "Error";
        Package2VersionStatus["initializing"] = "Initializing";
        Package2VersionStatus["verifyingFeaturesAndSettings"] = "VerifyingFeaturesAndSettings";
        Package2VersionStatus["verifyingDependencies"] = "VerifyingDependencies";
        Package2VersionStatus["verifyingMetadata"] = "VerifyingMetadata";
        Package2VersionStatus["finalizingPackageVersion"] = "FinalizingPackageVersion";
    })(Package2VersionStatus = PackagingSObjects.Package2VersionStatus || (PackagingSObjects.Package2VersionStatus = {}));
})(PackagingSObjects = exports.PackagingSObjects || (exports.PackagingSObjects = {}));
