// @salesforce/plugin-templates/lib/commands/static-resource/generate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const sf_plugins_core_1 = require("@salesforce/sf-plugins-core");
const staticResourceGenerator_1 = require("@salesforce/templates/lib/generators/staticResourceGenerator");
const core_1 = require("@salesforce/core");
const flags_1 = require("../../utils/flags");
const templateCommand_1 = require("../../utils/templateCommand");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-templates', 'staticResource');
class StaticResource extends sf_plugins_core_1.SfCommand {
    run() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { flags } = yield this.parse(StaticResource);
            // translate the new flags to the old ones the generator expects
            const flagsAsOptions = {
                resourcename: flags.name,
                contenttype: flags.type,
                template: 'empty',
                outputdir: flags['output-dir'],
                apiversion: flags['api-version'],
            };
            return (0, templateCommand_1.runGenerator)({
                generator: staticResourceGenerator_1.default,
                opts: flagsAsOptions,
                ux: new sf_plugins_core_1.Ux({ jsonEnabled: this.jsonEnabled() }),
                templates: (0, templateCommand_1.getCustomTemplates)(this.configAggregator),
            });
        });
    }
}
exports.default = StaticResource;
StaticResource.summary = messages.getMessage('summary');
StaticResource.description = messages.getMessage('description');
StaticResource.examples = messages.getMessages('examples');
StaticResource.aliases = ['force:staticresource:create'];
StaticResource.deprecateAliases = true;
StaticResource.flags = {
    name: sf_plugins_core_1.Flags.string({
        char: 'n',
        summary: messages.getMessage('flags.name.summary'),
        description: messages.getMessage('flags.name.description'),
        required: true,
        aliases: ['resourcename'],
        deprecateAliases: true,
    }),
    type: sf_plugins_core_1.Flags.string({
        summary: messages.getMessage('flags.type.summary'),
        description: messages.getMessage('flags.type.description'),
        default: 'application/zip',
        aliases: ['contenttype'],
        deprecateAliases: true,
    }),
    'output-dir': flags_1.outputDirFlag,
    'api-version': sf_plugins_core_1.orgApiVersionFlagWithDeprecations,
    loglevel: sf_plugins_core_1.loglevel,
};
