// @salesforce/plugin-packaging/lib/commands/package/delete.js
"use strict";
/*
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PackageDeleteCommand = void 0;
const sf_plugins_core_1 = require("@salesforce/sf-plugins-core");
const core_1 = require("@salesforce/core");
const packaging_1 = require("@salesforce/packaging");
const hubFlag_1 = require("../../utils/hubFlag");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-packaging', 'package_delete');
class PackageDeleteCommand extends sf_plugins_core_1.SfCommand {
    async run() {
        const { flags } = await this.parse(PackageDeleteCommand);
        const promptMsg = flags.undelete ? 'prompt-undelete' : 'prompt-delete';
        const accepted = flags['no-prompt'] || flags.json ? true : await this.confirm(messages.getMessage(promptMsg));
        if (!accepted) {
            throw messages.createError('prompt-delete-deny');
        }
        const pkg = new packaging_1.Package({
            connection: flags['target-dev-hub'].getConnection(flags['api-version']),
            project: this.project,
            packageAliasOrId: flags.package,
        });
        const result = flags.undelete ? await pkg.undelete() : await pkg.delete();
        this.display(result, flags.undelete);
        return result;
    }
    display(result, undelete) {
        this.log();
        this.logSuccess(messages.getMessage(undelete ? 'humanSuccessUndelete' : 'humanSuccess', [result.id]));
    }
}
exports.PackageDeleteCommand = PackageDeleteCommand;
PackageDeleteCommand.summary = messages.getMessage('summary');
PackageDeleteCommand.description = messages.getMessage('description');
PackageDeleteCommand.examples = messages.getMessages('examples');
PackageDeleteCommand.deprecateAliases = true;
PackageDeleteCommand.aliases = ['force:package:beta:delete', 'force:package:delete'];
PackageDeleteCommand.requiresProject = true;
PackageDeleteCommand.flags = {
    loglevel: sf_plugins_core_1.loglevel,
    'target-dev-hub': hubFlag_1.requiredHubFlag,
    'api-version': sf_plugins_core_1.orgApiVersionFlagWithDeprecations,
    'no-prompt': sf_plugins_core_1.Flags.boolean({
        char: 'n',
        deprecateAliases: true,
        aliases: ['noprompt'],
        summary: messages.getMessage('flags.no-prompt.summary'),
    }),
    package: sf_plugins_core_1.Flags.string({
        char: 'p',
        summary: messages.getMessage('flags.package.summary'),
        required: true,
    }),
    undelete: sf_plugins_core_1.Flags.boolean({
        summary: messages.getMessage('flags.undelete.summary'),
        hidden: true,
        default: false,
    }),
};
