// @salesforce/packaging/lib/package/packageVersionCreateRequest.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.byId = exports.list = void 0;
const util = require("util");
const core_1 = require("@salesforce/core");
const packageUtils_1 = require("../utils/packageUtils");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/packaging', 'package_version_create');
const STATUS_ERROR = 'Error';
const QUERY = 'SELECT Id, Status, Package2Id, Package2VersionId, Package2Version.SubscriberPackageVersionId, Tag, Branch, ' +
    'CreatedDate, Package2Version.HasMetadataRemoved, CreatedById ' +
    'FROM Package2VersionCreateRequest ' +
    '%s' + // WHERE, if applicable
    'ORDER BY CreatedDate desc';
function formatDate(date) {
    const pad = (num) => (num < 10 ? `0${num}` : `${num}`);
    return `${date.getFullYear()}-${pad(date.getMonth() + 1)}-${pad(date.getDate())} ${pad(date.getHours())}:${pad(date.getMinutes())}`;
}
async function list(options) {
    try {
        const whereClause = constructWhere(options);
        return await query(util.format(QUERY, whereClause), options.connection);
    }
    catch (err) {
        throw (0, packageUtils_1.applyErrorAction)((0, packageUtils_1.massageErrorMessage)(err));
    }
}
exports.list = list;
async function byId(packageVersionCreateRequestId, connection) {
    const results = await query(util.format(QUERY, `WHERE Id = '${packageVersionCreateRequestId}' `), connection);
    if (results && results.length === 1 && results[0].Status === STATUS_ERROR) {
        results[0].Error = await queryForErrors(packageVersionCreateRequestId, connection);
    }
    return results;
}
exports.byId = byId;
// eslint-disable-next-line @typescript-eslint/no-shadow
async function query(query, connection) {
    const queryResult = await connection.autoFetchQuery(query, { tooling: true });
    return (queryResult.records ? queryResult.records : []).map((record) => ({
        Id: record.Id,
        Status: record.Status,
        Package2Id: record.Package2Id,
        Package2VersionId: record.Package2VersionId,
        SubscriberPackageVersionId: record.Package2Version != null ? record.Package2Version.SubscriberPackageVersionId : null,
        Tag: record.Tag,
        Branch: record.Branch,
        Error: [],
        CreatedDate: formatDate(new Date(record.CreatedDate)),
        HasMetadataRemoved: record.Package2Version != null ? record.Package2Version.HasMetadataRemoved : null,
        CreatedBy: record.CreatedById,
    }));
}
async function queryForErrors(packageVersionCreateRequestId, connection) {
    const errorResults = [];
    const queryResult = await connection.tooling.query(util.format("SELECT Message FROM Package2VersionCreateRequestError WHERE ParentRequest.Id = '%s'", packageVersionCreateRequestId));
    if (queryResult.records) {
        queryResult.records.forEach((record) => {
            errorResults.push(record.Message);
        });
    }
    return errorResults;
}
function constructWhere(options) {
    const where = [];
    if (options?.id) {
        where.push(`Id = '${options.id}'`);
    }
    // filter on created date, days ago: 0 for today, etc
    if (options?.createdlastdays) {
        if (options.createdlastdays < 0) {
            throw messages.createError('invalidDaysNumber', ['createdlastdays', options.createdlastdays]);
        }
        where.push(`CreatedDate = LAST_N_DAYS:${options.createdlastdays}`);
    }
    // filter on errors
    if (options?.status) {
        where.push(`Status = '${options.status.toLowerCase()}'`);
    }
    return where.length > 0 ? `WHERE ${where.join(' AND ')}` : '';
}
