// @salesforce/plugin-templates/lib/commands/apex/generate/class.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const sf_plugins_core_1 = require("@salesforce/sf-plugins-core");
const apexClassGenerator_1 = require("@salesforce/templates/lib/generators/apexClassGenerator");
const utils_1 = require("@salesforce/templates/lib/utils");
const core_1 = require("@salesforce/core");
const templateCommand_1 = require("../../../utils/templateCommand");
const flags_1 = require("../../../utils/flags");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-templates', 'apexClass');
const commonMessages = core_1.Messages.loadMessages('@salesforce/plugin-templates', 'messages');
const apexClassFileSuffix = /.cls$/;
class ApexClass extends sf_plugins_core_1.SfCommand {
    run() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { flags } = yield this.parse(ApexClass);
            const flagsAsOptions = {
                apiversion: flags['api-version'],
                classname: flags.name,
                template: flags.template,
                outputdir: flags['output-dir'],
            };
            return (0, templateCommand_1.runGenerator)({
                generator: apexClassGenerator_1.default,
                opts: flagsAsOptions,
                ux: new sf_plugins_core_1.Ux({ jsonEnabled: this.jsonEnabled() }),
                templates: (0, templateCommand_1.getCustomTemplates)(this.configAggregator),
            });
        });
    }
}
exports.default = ApexClass;
ApexClass.summary = messages.getMessage('summary');
ApexClass.description = messages.getMessage('description');
ApexClass.examples = messages.getMessages('examples');
ApexClass.aliases = ['force:apex:class:create'];
ApexClass.deprecateAliases = true;
ApexClass.flags = {
    name: sf_plugins_core_1.Flags.string({
        char: 'n',
        summary: messages.getMessage('flags.name.summary'),
        description: messages.getMessage('flags.name.description'),
        required: true,
        aliases: ['classname'],
        deprecateAliases: true,
    }),
    template: sf_plugins_core_1.Flags.string({
        char: 't',
        summary: commonMessages.getMessage('flags.template'),
        description: commonMessages.getMessage('flags.template.description'),
        default: 'DefaultApexClass',
        options: utils_1.CreateUtil.getCommandTemplatesForFiletype(apexClassFileSuffix, 'apexclass'),
    }),
    'output-dir': flags_1.outputDirFlag,
    'api-version': sf_plugins_core_1.orgApiVersionFlagWithDeprecations,
    loglevel: sf_plugins_core_1.loglevel,
};
