// @salesforce/plugin-user/lib/baseCommands/user/permsetlicense/assign.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserPermSetLicenseAssignBaseCommand = void 0;
const core_1 = require("@salesforce/core");
const sf_plugins_core_1 = require("@salesforce/sf-plugins-core");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-user', 'permsetlicense.assign');
class UserPermSetLicenseAssignBaseCommand extends sf_plugins_core_1.SfCommand {
    constructor() {
        super(...arguments);
        this.usernamesOrAliases = [];
        this.successes = [];
        this.failures = [];
    }
    async assign() {
        this.logger = await core_1.Logger.child(this.constructor.name);
        this.logger.debug(`will assign perm set license "${this.pslName}" to users: ${this.usernamesOrAliases.join(', ')}`);
        try {
            this.pslId = (await this.connection.singleRecordQuery(`select Id from PermissionSetLicense where DeveloperName = '${this.pslName}' or MasterLabel = '${this.pslName}'`)).Id;
        }
        catch {
            throw new core_1.SfError('PermissionSetLicense not found');
        }
        (await Promise.all(this.usernamesOrAliases.map((usernameOrAlias) => this.usernameToPSLAssignment({
            pslName: this.pslName,
            usernameOrAlias,
        })))).map((result) => {
            if (isSuccess(result)) {
                this.successes.push(result);
            }
            else {
                this.failures.push(result);
            }
        });
        this.print();
        this.setExitCode();
        return {
            successes: this.successes,
            failures: this.failures,
        };
    }
    // handles one username/psl combo so these can run in parallel
    async usernameToPSLAssignment({ pslName, usernameOrAlias, }) {
        // Convert any aliases to usernames
        const resolvedUsername = (await core_1.StateAggregator.getInstance()).aliases.resolveUsername(usernameOrAlias);
        try {
            const AssigneeId = (await this.connection.singleRecordQuery(`select Id from User where Username = '${resolvedUsername}'`)).Id;
            await this.connection.sobject('PermissionSetLicenseAssign').create({
                AssigneeId,
                PermissionSetLicenseId: this.pslId,
            });
            return {
                name: resolvedUsername,
                value: pslName,
            };
        }
        catch (e) {
            // idempotency.  If user(s) already have PSL, the API will throw an error about duplicate value.
            // but we're going to call that a success
            if (e instanceof Error && e.message.startsWith('duplicate value found')) {
                this.warn(messages.getMessage('duplicateValue', [resolvedUsername, pslName]));
                return {
                    name: resolvedUsername,
                    value: pslName,
                };
            }
            else {
                return {
                    name: resolvedUsername,
                    message: e instanceof Error ? e.message : 'error contained no message',
                };
            }
        }
    }
    setExitCode() {
        if (this.failures.length && this.successes.length) {
            process.exitCode = 68;
        }
        else if (this.failures.length) {
            process.exitCode = 1;
        }
        else if (this.successes.length) {
            process.exitCode = 0;
        }
    }
    print() {
        if (this.failures.length > 0 && this.successes.length > 0) {
            this.styledHeader('Partial Success');
        }
        if (this.successes.length > 0) {
            this.styledHeader('Permset Licenses Assigned');
            this.table(this.successes, {
                name: { header: 'Username' },
                value: { header: 'Permission Set License Assignment' },
            });
        }
        if (this.failures.length > 0) {
            if (this.successes.length > 0) {
                this.log('');
            }
            this.styledHeader('Failures');
            this.table(this.failures, { name: { header: 'Username' } }, { message: { header: 'Error Message' } });
        }
    }
}
exports.UserPermSetLicenseAssignBaseCommand = UserPermSetLicenseAssignBaseCommand;
const isSuccess = (input) => input.value !== undefined;
