// @salesforce/source-deploy-retrieve/lib/src/convert/metadataConverter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetadataConverter = void 0;
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const stream_1 = require("stream");
const path_1 = require("path");
const core_1 = require("@salesforce/core");
const graceful_fs_1 = require("graceful-fs");
const ts_types_1 = require("@salesforce/ts-types");
const fileSystemHandler_1 = require("../utils/fileSystemHandler");
const collections_1 = require("../collections");
const registry_1 = require("../registry");
const streams_1 = require("./streams");
const replacements_1 = require("./replacements");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/source-deploy-retrieve', 'sdr');
class MetadataConverter {
    constructor(registry = new registry_1.RegistryAccess()) {
        this.registry = registry;
    }
    // eslint-disable-next-line complexity
    async convert(comps, targetFormat, output) {
        try {
            const cs = comps instanceof collections_1.ComponentSet ? comps : new collections_1.ComponentSet(comps, this.registry);
            const components = (comps instanceof collections_1.ComponentSet ? Array.from(comps.getSourceComponents()) : comps).filter((comp) => comp.type.isAddressable !== false);
            if (output.type !== 'merge' && output.packageName) {
                cs.fullName = output.packageName;
            }
            const targetFormatIsSource = targetFormat === 'source';
            const { packagePath, defaultDirectory, writer, mergeSet, tasks = [], } = await getConvertIngredients(output, cs, targetFormatIsSource);
            const conversionPipeline = (0, streams_1.pipeline)(stream_1.Readable.from(components), !targetFormatIsSource && (process.env.SF_APPLY_REPLACEMENTS_ON_CONVERT === 'true' || output.type === 'zip')
                ? (await (0, replacements_1.getReplacementMarkingStream)(cs.projectDirectory)) ?? new stream_1.PassThrough({ objectMode: true })
                : new stream_1.PassThrough({ objectMode: true }), new streams_1.ComponentConverter(targetFormat, this.registry, mergeSet, defaultDirectory), writer);
            await Promise.all([conversionPipeline, ...tasks]);
            const result = { packagePath };
            if (output.type === 'zip' && !packagePath) {
                result.zipBuffer = writer.buffer;
            }
            else if (output.type !== 'zip') {
                result.converted = writer.converted;
            }
            return result;
        }
        catch (err) {
            if (!(err instanceof Error) && !(0, ts_types_1.isString)(err)) {
                throw err;
            }
            // if the error is already somewhat descriptive, use that
            // the allows better error messages to be passed through instead of "failed convert"
            if (err instanceof core_1.SfError && (err.name !== 'SfError' || err.actions)) {
                throw err;
            }
            const error = (0, ts_types_1.isString)(err) ? new Error(err) : err;
            throw new core_1.SfError(messages.getMessage('error_failed_convert', [error.message]), 'ConversionError', [], error);
        }
    }
}
exports.MetadataConverter = MetadataConverter;
MetadataConverter.PACKAGE_XML_FILE = 'package.xml';
MetadataConverter.DESTRUCTIVE_CHANGES_POST_XML_FILE = 'destructiveChangesPost.xml';
MetadataConverter.DESTRUCTIVE_CHANGES_PRE_XML_FILE = 'destructiveChangesPre.xml';
MetadataConverter.DEFAULT_PACKAGE_PREFIX = 'metadataPackage';
function getPackagePath(outputConfig) {
    let packagePath;
    const { genUniqueDir = true, outputDirectory, packageName, type } = outputConfig;
    if (outputDirectory) {
        if (packageName) {
            packagePath = (0, path_1.join)(outputDirectory, packageName);
        }
        else if (genUniqueDir) {
            packagePath = (0, path_1.join)(outputDirectory, `${MetadataConverter.DEFAULT_PACKAGE_PREFIX}_${Date.now()}`);
        }
        else {
            packagePath = (0, path_1.normalize)(outputDirectory);
        }
        if (type === 'zip') {
            packagePath += '.zip';
            (0, fileSystemHandler_1.ensureDirectoryExists)((0, path_1.dirname)(packagePath));
        }
        else {
            (0, fileSystemHandler_1.ensureDirectoryExists)(packagePath);
        }
    }
    return packagePath;
}
const getDestructiveManifest = (destructiveChangesType) => {
    switch (destructiveChangesType) {
        case collections_1.DestructiveChangesType.POST:
            return MetadataConverter.DESTRUCTIVE_CHANGES_POST_XML_FILE;
        case collections_1.DestructiveChangesType.PRE:
            return MetadataConverter.DESTRUCTIVE_CHANGES_PRE_XML_FILE;
    }
};
async function getConvertIngredients(output, cs, targetFormatIsSource) {
    switch (output.type) {
        case 'directory':
            return getDirectoryConfigOutputs(output, targetFormatIsSource, cs);
        case 'zip':
            return getZipConfigOutputs(output, targetFormatIsSource, cs);
        case 'merge':
            return getMergeConfigOutputs(output, targetFormatIsSource);
    }
}
function getMergeConfigOutputs(output, targetFormatIsSource) {
    if (!targetFormatIsSource) {
        throw new core_1.SfError(messages.getMessage('error_merge_metadata_target_unsupported'));
    }
    const defaultDirectory = output.defaultDirectory;
    const mergeSet = new collections_1.ComponentSet();
    // since child components are composed in metadata format, we need to merge using the parent
    for (const component of output.mergeWith) {
        mergeSet.add(component.parent ?? component);
    }
    const writer = new streams_1.StandardWriter(output.defaultDirectory);
    if (output.forceIgnoredPaths) {
        writer.forceIgnoredPaths = output.forceIgnoredPaths;
    }
    return {
        writer,
        mergeSet,
        defaultDirectory,
    };
}
async function getZipConfigOutputs(output, targetFormatIsSource, cs) {
    const packagePath = getPackagePath(output);
    const writer = new streams_1.ZipWriter(packagePath);
    if (!targetFormatIsSource) {
        writer.addToZip(await cs.getPackageXml(), MetadataConverter.PACKAGE_XML_FILE);
        // for each of the destructive changes in the component set, convert and write the correct metadata to each manifest
        await Promise.all(cs
            .getTypesOfDestructiveChanges()
            .map(async (destructiveChangeType) => writer.addToZip(await cs.getPackageXml(4, destructiveChangeType), getDestructiveManifest(destructiveChangeType))));
    }
    return {
        packagePath,
        defaultDirectory: packagePath,
        writer,
        mergeSet: undefined,
    };
}
async function getDirectoryConfigOutputs(output, targetFormatIsSource, cs) {
    const packagePath = getPackagePath(output);
    return {
        packagePath,
        defaultDirectory: packagePath,
        writer: new streams_1.StandardWriter(packagePath),
        tasks: targetFormatIsSource
            ? []
            : [
                graceful_fs_1.promises.writeFile((0, path_1.join)(packagePath, MetadataConverter.PACKAGE_XML_FILE), await cs.getPackageXml()),
                ...cs.getTypesOfDestructiveChanges().map(async (destructiveChangesType) => 
                // for each of the destructive changes in the component set, convert and write the correct metadata
                // to each manifest
                graceful_fs_1.promises.writeFile((0, path_1.join)(packagePath, getDestructiveManifest(destructiveChangesType)), await cs.getPackageXml(4, destructiveChangesType))),
            ],
    };
}
