// @salesforce/plugin-trust/lib/commands/plugins/trust/verify.js
"use strict";
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Verify = void 0;
const sf_plugins_core_1 = require("@salesforce/sf-plugins-core");
const core_1 = require("@salesforce/core");
const installationVerification_1 = require("../../../shared/installationVerification");
const NpmName_1 = require("../../../shared/NpmName");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-trust', 'verify');
class Verify extends sf_plugins_core_1.SfCommand {
    static getVerifier(npmName, config) {
        return new installationVerification_1.InstallationVerification().setPluginNpmName(npmName).setConfig(config);
    }
    async run() {
        const { flags } = await this.parse(Verify);
        const logger = await core_1.Logger.child('verify');
        this.log('Checking for digital signature.');
        const npmName = NpmName_1.NpmName.parse(flags.npm);
        logger.debug(`running verify command for npm: ${npmName.name}`);
        const vConfig = new installationVerification_1.VerificationConfig();
        const configContext = {
            cacheDir: this.config.cacheDir,
            configDir: this.config.configDir,
            dataDir: this.config.dataDir,
            cliRoot: this.config.root,
        };
        logger.debug(`cacheDir: ${configContext.cacheDir}`);
        logger.debug(`configDir: ${configContext.configDir}`);
        logger.debug(`dataDir: ${configContext.dataDir}`);
        vConfig.verifier = Verify.getVerifier(npmName, configContext);
        if (flags.registry) {
            process.env.SFDX_NPM_REGISTRY = flags.registry;
        }
        try {
            const meta = await vConfig.verifier.verify();
            logger.debug(`meta.verified: ${meta.verified}`);
            if (!meta.verified) {
                const e = messages.createError('FailedDigitalSignatureVerification');
                // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                // @ts-ignore override readonly .name field
                e.name = 'FailedDigitalSignatureVerification';
                throw e;
            }
            const message = `Successfully validated digital signature for ${npmName.name}.`;
            if (!flags.json) {
                vConfig.log(message);
            }
            else {
                return { message, verified: true };
            }
        }
        catch (error) {
            const err = error;
            logger.debug(`err reported: ${JSON.stringify(err, null, 4)}`);
            const response = {
                verified: false,
                message: err.message,
            };
            if (err.name === 'NotSigned') {
                let message = err.message;
                if (await vConfig.verifier.isAllowListed()) {
                    message = `The plugin [${npmName.name}] is not digitally signed but it is allow-listed.`;
                    vConfig.log(message);
                    response.message = message;
                }
                else {
                    message = 'The plugin is not digitally signed.';
                    vConfig.log(message);
                    response.message = message;
                }
                return response;
            }
            throw core_1.SfError.wrap(err);
        }
    }
}
exports.Verify = Verify;
Verify.summary = messages.getMessage('summary');
Verify.description = messages.getMessage('description');
Verify.examples = messages.getMessages('examples');
Verify.flags = {
    npm: sf_plugins_core_1.Flags.string({
        char: 'n',
        required: true,
        summary: messages.getMessage('flags.npm'),
    }),
    registry: sf_plugins_core_1.Flags.string({
        char: 'r',
        summary: messages.getMessage('flags.registry'),
    }),
    loglevel: sf_plugins_core_1.loglevel,
};
