// @salesforce/plugin-community/lib/shared/community/connect/CommunityTemplatesResource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommunityTemplatesResource = void 0;
const core_1 = require("@salesforce/core");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-community', 'template.list');
/**
 * A connect api resource for fetching community templates available to context user
 */
class CommunityTemplatesResource {
    constructor(ux) {
        this.ux = ux;
    }
    // eslint-disable-next-line @typescript-eslint/require-await, class-methods-use-this
    async fetchRelativeConnectUrl() {
        return '/connect/communities/templates';
    }
    // eslint-disable-next-line @typescript-eslint/require-await, class-methods-use-this
    async fetchPostParams() {
        return JSON.stringify({});
    }
    // eslint-disable-next-line class-methods-use-this
    getRequestMethod() {
        return 'GET';
    }
    handleSuccess(result) {
        const columns = {
            templateName: { header: 'Template Name' },
            publisher: { header: 'Publisher' },
        };
        this.ux.styledHeader(messages.getMessage('response.styledHeader'));
        this.ux.table(result.templates, columns);
        this.ux.log();
        this.ux.log(messages.getMessage('response.TotalField'), result.total.toString());
        return result;
    }
    // eslint-disable-next-line class-methods-use-this
    handleError(error) {
        throw error;
    }
}
exports.CommunityTemplatesResource = CommunityTemplatesResource;
