// @salesforce/plugin-packaging/lib/commands/package/version/create/report.js
"use strict";
/*
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PackageVersionCreateReportCommand = void 0;
const sf_plugins_core_1 = require("@salesforce/sf-plugins-core");
const core_1 = require("@salesforce/core");
const pkgUtils = require("@salesforce/packaging");
const packaging_1 = require("@salesforce/packaging");
const chalk = require("chalk");
const kit_1 = require("@salesforce/kit");
const hubFlag_1 = require("../../../../utils/hubFlag");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-packaging', 'package_version_create_report');
const pvclMessages = core_1.Messages.loadMessages('@salesforce/plugin-packaging', 'package_version_create_list');
const plMessages = core_1.Messages.loadMessages('@salesforce/plugin-packaging', 'package_list');
const ERROR_LIMIT = 12;
class PackageVersionCreateReportCommand extends sf_plugins_core_1.SfCommand {
    async run() {
        const { flags } = await this.parse(PackageVersionCreateReportCommand);
        const result = await packaging_1.PackageVersion.getCreateStatus(flags['package-create-request-id'], flags['target-dev-hub'].getConnection(flags['api-version']));
        this.display(result, flags['package-create-request-id'], flags['target-dev-hub']);
        return [result];
    }
    display(record, requestId, devOrg) {
        const installUrlValue = record.Status === 'Success' ? `${pkgUtils.INSTALL_URL_BASE.toString()}${record.SubscriberPackageVersionId}` : '';
        const data = [
            {
                key: pvclMessages.getMessage('id'),
                value: record.Id,
            },
            {
                key: pvclMessages.getMessage('status'),
                value: (0, kit_1.camelCaseToTitleCase)(record.Status),
            },
            {
                key: pvclMessages.getMessage('package-id'),
                value: record.Package2Id,
            },
            {
                key: pvclMessages.getMessage('packageVersionId'),
                value: record.Package2VersionId,
            },
            {
                key: pvclMessages.getMessage('subscriberPackageVersionId'),
                value: record.SubscriberPackageVersionId,
            },
            {
                key: pvclMessages.getMessage('tag'),
                value: record.Tag,
            },
            {
                key: pvclMessages.getMessage('branch'),
                value: record.Branch,
            },
            { key: 'Created Date', value: record.CreatedDate },
            {
                key: pvclMessages.getMessage('installUrl'),
                value: installUrlValue,
            },
            {
                key: plMessages.getMessage('createdBy'),
                value: record.CreatedBy,
            },
        ];
        this.styledHeader(chalk.blue('Package Version Create Request'));
        this.table(data, {
            key: { header: 'Name' },
            value: { header: 'Value' },
        });
        if (record.Error?.length > 0) {
            this.log('');
            const errors = [];
            record.Error.slice(0, ERROR_LIMIT).forEach((error) => {
                errors.push(`(${errors.length + 1}) ${error}`);
            });
            this.styledHeader(chalk.red('Errors'));
            this.warn(errors.join('\n'));
            // Check if errors were truncated.  If so, inform the user with
            // instructions on how to retrieve the remaining errors.
            if (record.Error.length > ERROR_LIMIT) {
                this.warn(messages.getMessage('truncatedErrors', [this.config.bin, requestId, devOrg.getUsername()]));
            }
        }
    }
}
exports.PackageVersionCreateReportCommand = PackageVersionCreateReportCommand;
PackageVersionCreateReportCommand.summary = messages.getMessage('summary');
PackageVersionCreateReportCommand.description = messages.getMessage('description');
PackageVersionCreateReportCommand.examples = messages.getMessages('examples');
PackageVersionCreateReportCommand.deprecateAliases = true;
PackageVersionCreateReportCommand.aliases = ['force:package:beta:version:create:report', 'force:package:version:create:report'];
PackageVersionCreateReportCommand.flags = {
    loglevel: sf_plugins_core_1.loglevel,
    'target-dev-hub': hubFlag_1.requiredHubFlag,
    'api-version': sf_plugins_core_1.orgApiVersionFlagWithDeprecations,
    // eslint-disable-next-line sf-plugin/id-flag-suggestions
    'package-create-request-id': sf_plugins_core_1.Flags.salesforceId({
        length: 'both',
        deprecateAliases: true,
        aliases: ['packagecreaterequestid'],
        char: 'i',
        summary: messages.getMessage('flags.package-create-request-id.summary'),
        required: true,
    }),
};
