// @salesforce/plugin-templates/lib/commands/analytics/generate/template.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const sf_plugins_core_1 = require("@salesforce/sf-plugins-core");
const core_1 = require("@salesforce/core");
const analyticsTemplateGenerator_1 = require("@salesforce/templates/lib/generators/analyticsTemplateGenerator");
const templateCommand_1 = require("../../../utils/templateCommand");
const flags_1 = require("../../../utils/flags");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-templates', 'analyticsTemplate');
class AnalyticsTemplate extends sf_plugins_core_1.SfCommand {
    run() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { flags } = yield this.parse(AnalyticsTemplate);
            const flagsAsOptions = {
                apiversion: flags['api-version'],
                templatename: flags.name,
                outputdir: flags['output-dir'],
            };
            return (0, templateCommand_1.runGenerator)({
                generator: analyticsTemplateGenerator_1.default,
                opts: flagsAsOptions,
                ux: new sf_plugins_core_1.Ux({ jsonEnabled: this.jsonEnabled() }),
                templates: (0, templateCommand_1.getCustomTemplates)(this.configAggregator),
            });
        });
    }
}
exports.default = AnalyticsTemplate;
AnalyticsTemplate.examples = messages.getMessages('examples');
AnalyticsTemplate.summary = messages.getMessage('summary');
AnalyticsTemplate.description = messages.getMessage('description');
AnalyticsTemplate.aliases = ['force:analytics:template:create'];
AnalyticsTemplate.deprecateAliases = true;
AnalyticsTemplate.flags = {
    'output-dir': flags_1.outputDirFlag,
    'api-version': sf_plugins_core_1.orgApiVersionFlagWithDeprecations,
    name: sf_plugins_core_1.Flags.string({
        char: 'n',
        summary: messages.getMessage('flags.name.summary'),
        required: true,
        aliases: ['templatename'],
        deprecateAliases: true,
    }),
    loglevel: sf_plugins_core_1.loglevel,
};
