// @salesforce/plugin-packaging/lib/commands/package1/version/display.js
"use strict";
/*
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Package1VersionDisplayCommand = void 0;
const sf_plugins_core_1 = require("@salesforce/sf-plugins-core");
const core_1 = require("@salesforce/core");
const packaging_1 = require("@salesforce/packaging");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-packaging', 'package1_version_display');
class Package1VersionDisplayCommand extends sf_plugins_core_1.SfCommand {
    async run() {
        const { flags } = await this.parse(Package1VersionDisplayCommand);
        const pv1 = new packaging_1.Package1Version(flags['target-org'].getConnection(flags['api-version']), flags['package-version-id']);
        const results = (await pv1.getPackageVersion()).map((result) => ({
            MetadataPackageVersionId: result.Id,
            MetadataPackageId: result.MetadataPackageId,
            Name: result.Name,
            ReleaseState: result.ReleaseState,
            Version: `${result.MajorVersion}.${result.MinorVersion}.${result.PatchVersion}`,
            BuildNumber: result.BuildNumber,
        }));
        if (results.length === 0) {
            this.warn('No results found');
        }
        else {
            this.table(results, {
                MetadataPackageVersionId: { header: 'MetadataPackageVersionId' },
                MetadataPackageId: { header: 'MetadataPackageId' },
                Name: { header: 'Name' },
                Version: { header: 'Version' },
                ReleaseState: { header: 'ReleaseState' },
                BuildNumber: { header: 'BuildNumber' },
            });
        }
        return results;
    }
}
exports.Package1VersionDisplayCommand = Package1VersionDisplayCommand;
Package1VersionDisplayCommand.summary = messages.getMessage('summary');
Package1VersionDisplayCommand.examples = messages.getMessages('examples');
Package1VersionDisplayCommand.deprecateAliases = true;
Package1VersionDisplayCommand.aliases = ['force:package1:beta:version:display', 'force:package1:version:display'];
Package1VersionDisplayCommand.flags = {
    loglevel: sf_plugins_core_1.loglevel,
    'target-org': sf_plugins_core_1.requiredOrgFlagWithDeprecations,
    'api-version': sf_plugins_core_1.orgApiVersionFlagWithDeprecations,
    'package-version-id': sf_plugins_core_1.Flags.salesforceId({
        length: 'both',
        char: 'i',
        deprecateAliases: true,
        aliases: ['packageversionid'],
        summary: messages.getMessage('flags.package-version-id.summary'),
        required: true,
        startsWith: '04t',
    }),
};
