// @salesforce/plugin-community/lib/shared/connect/services/ConnectExecutor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConnectExecutor = void 0;
const core_1 = require("@salesforce/core");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-community', 'connect-executor');
/**
 * An executor which calls a connect api for the given org
 */
class ConnectExecutor {
    constructor(connectService, org) {
        this.connectService = connectService;
        this.org = org;
    }
    /**
     * Call the connect resource as defined by the given ConnectResource for the given org
     */
    async callConnectApi() {
        return this.org
            .getConnection()
            .request(await this.fetchRequestInfo())
            .then((result) => this.connectService.handleSuccess(result))
            .catch((err) => this.connectService.handleError(err));
    }
    async fetchRequestInfo() {
        const connectUrl = encodeURI(await this.connectService.fetchRelativeConnectUrl());
        const method = this.connectService.getRequestMethod();
        if (method === 'GET') {
            return {
                url: connectUrl,
                method,
                body: null,
            };
        }
        else if (method === 'POST') {
            return {
                url: connectUrl,
                method,
                body: await this.connectService.fetchPostParams(),
            };
        }
        else {
            throw new core_1.SfError(messages.getMessage('unsupportedOperation', [method]), 'UNSUPPORTED_OPERATION');
        }
    }
}
exports.ConnectExecutor = ConnectExecutor;
