// @salesforce/plugin-source/lib/formatters/source/pushResultFormatter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mergeReplacements = exports.PushResultFormatter = void 0;
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const path_1 = require("path");
const chalk = require("chalk");
const core_1 = require("@salesforce/core");
const source_deploy_retrieve_1 = require("@salesforce/source-deploy-retrieve");
const ts_types_1 = require("@salesforce/ts-types");
const kit_1 = require("@salesforce/kit");
const resultFormatter_1 = require("../resultFormatter");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-source', 'push');
class PushResultFormatter extends resultFormatter_1.ResultFormatter {
    constructor(logger, ux, options, results, 
    // if your push included deletes that are bundle subcomponents, we'll need to add those deletes to the results even though they aren't included in fileResponses
    deletes = []) {
        super(logger, ux, options);
        this.results = results;
        this.deletes = deletes;
        this.fileResponses = this.correctFileResponses();
        this.replacements = (0, exports.mergeReplacements)(results);
    }
    /**
     * Get the JSON output from the DeployResult.
     *
     * @returns a JSON formatted result matching the provided type.
     */
    getJson() {
        // throws a particular json structure.  commandName property will be appended by sfdxCommand when this throws
        if (process.exitCode !== 0) {
            const error = new core_1.SfError(messages.getMessage('sourcepushFailed', ['']), 'DeployFailed', [], process.exitCode);
            const errorData = this.fileResponses.filter((fileResponse) => fileResponse.state === source_deploy_retrieve_1.ComponentStatus.Failed);
            error.setData(errorData);
            error['result'] = errorData;
            // partial success
            if (process.exitCode === 69) {
                error['partialSuccess'] = this.fileResponses.filter((fileResponse) => fileResponse.state !== source_deploy_retrieve_1.ComponentStatus.Failed);
            }
            throw error;
        }
        // quiet returns only failures
        const toReturn = this.isQuiet()
            ? this.fileResponses.filter((fileResponse) => fileResponse.state === source_deploy_retrieve_1.ComponentStatus.Failed)
            : this.fileResponses;
        return {
            pushedSource: toReturn.map(({ state, fullName, type, filePath }) => ({ state, fullName, type, filePath })),
            ...(!this.isQuiet() && this.replacements.size ? { replacements: Object.fromEntries(this.replacements) } : {}),
        };
    }
    /**
     * Displays deploy results in human readable format.  Output can vary based on:
     *
     * 1. Verbose option
     * 3. Checkonly deploy (checkonly=true)
     * 4. Deploy with test results
     * 5. Canceled status
     */
    display() {
        this.displaySuccesses();
        this.displayFailures();
        this.displayReplacements();
        // Throw a DeployFailed error unless the deployment was successful.
        if (!this.isSuccess()) {
            // Add error message directly on the DeployResult (e.g., a GACK)
            let errMsg = '';
            this.results?.forEach((res) => {
                if (res.response?.errorMessage) {
                    errMsg += `${res.response?.errorMessage}\n`;
                }
            });
            throw new core_1.SfError(messages.getMessage('sourcepushFailed', [errMsg]), 'PushFailed');
        }
    }
    correctFileResponses() {
        const withoutUnchanged = this.results.some((result) => result.getFileResponses().length)
            ? this.results.flatMap((result) => result.getFileResponses().filter((fileResponse) => fileResponse.state !== 'Unchanged'))
            : [];
        if (!this.deletes.length) {
            return withoutUnchanged;
        }
        const bundlesDeployed = withoutUnchanged.filter((fileResponse) => ['LightningComponentBundle', 'AuraDefinitionBundle', 'WaveTemplateBundle'].includes(fileResponse.type));
        if (bundlesDeployed.length === 0) {
            return withoutUnchanged;
        }
        // "content" property of the bundles as a string
        const contentFilePathFromDeployedBundles = this.componentsFromFilenames(bundlesDeployed.map((fileResponse) => fileResponse.filePath))
            .map((c) => c.content)
            .filter(ts_types_1.isString);
        // there may be deletes not represented in the file responses (if bundle type)
        const resolver = new source_deploy_retrieve_1.MetadataResolver(undefined, source_deploy_retrieve_1.VirtualTreeContainer.fromFilePaths(this.deletes));
        return this.deletes
            .map((filePath) => {
            const cmp = this.resolveComponentsOrWarn(filePath, resolver)[0];
            if (cmp instanceof source_deploy_retrieve_1.SourceComponent &&
                cmp.type.strategies?.adapter === 'bundle' &&
                contentFilePathFromDeployedBundles.includes((0, path_1.resolve)(cmp.content))) {
                return {
                    state: source_deploy_retrieve_1.ComponentStatus.Deleted,
                    fullName: cmp.fullName,
                    type: cmp.type.name,
                    filePath,
                };
            }
        })
            .filter((fileResponse) => fileResponse)
            .concat(withoutUnchanged);
    }
    displaySuccesses() {
        if (this.isQuiet()) {
            return;
        }
        if (this.isSuccess() && this.fileResponses?.length) {
            const successes = this.fileResponses.filter((f) => f.state !== 'Failed');
            if (!successes.length) {
                return;
            }
            this.sortFileResponses(successes);
            this.asRelativePaths(successes);
            this.ux.log('');
            this.ux.styledHeader(chalk.blue('Pushed Source'));
            this.ux.table(successes, {
                state: { header: 'STATE' },
                fullName: { header: 'FULL NAME' },
                type: { header: 'TYPE' },
                filePath: { header: 'PROJECT PATH' },
            });
        }
    }
    displayReplacements() {
        if (!this.isQuiet() && this.replacements.size) {
            this.ux.log('');
            this.ux.styledHeader(chalk.blue('Metadata Replacements'));
            const replacements = Array.from(this.replacements.entries()).flatMap(([filepath, stringsReplaced]) => stringsReplaced.map((replaced) => ({
                filePath: (0, path_1.relative)(process.cwd(), filepath),
                replaced,
            })));
            this.ux.table(replacements, {
                filePath: { header: 'PROJECT PATH' },
                replaced: { header: 'TEXT REPLACED' },
            });
        }
    }
    displayFailures() {
        const failures = [];
        const fileResponseFailures = new Map();
        if (this.fileResponses?.length) {
            const fileResponses = [];
            this.fileResponses
                .filter((f) => f.state === 'Failed')
                .map((f) => {
                fileResponses.push(f);
                fileResponseFailures.set(`${f.type}#${f.fullName}`, f.error);
            });
            this.sortFileResponses(fileResponses);
            this.asRelativePaths(fileResponses);
            failures.push(...fileResponses);
        }
        const deployMessages = this.results?.flatMap((result) => (0, kit_1.ensureArray)(result.response?.details?.componentFailures));
        if (deployMessages.length > failures.length) {
            // if there's additional failures in the API response, find the failure and add it to the output
            deployMessages.map((deployMessage) => {
                if (!fileResponseFailures.has(`${deployMessage.componentType}#${deployMessage.fullName}`)) {
                    // duplicate the problem message to the error property for displaying in the table
                    failures.push(Object.assign(deployMessage, { error: deployMessage.problem }));
                }
            });
        }
        if (!failures.length) {
            return;
        }
        this.ux.log('');
        this.ux.styledHeader(chalk.red(`Component Failures [${failures.length}]`));
        this.ux.table(failures, {
            problemType: { header: 'Type' },
            fullName: { header: 'Name' },
            error: { header: 'Problem' },
        });
        this.ux.log('');
    }
    componentsFromFilenames(filenames) {
        const resolver = new source_deploy_retrieve_1.MetadataResolver(undefined, source_deploy_retrieve_1.VirtualTreeContainer.fromFilePaths(filenames));
        return filenames
            .flatMap((filename) => this.resolveComponentsOrWarn(filename, resolver))
            .filter((cmp) => cmp instanceof source_deploy_retrieve_1.SourceComponent);
    }
    resolveComponentsOrWarn(filename, resolver) {
        try {
            return resolver.getComponentsFromPath(filename);
        }
        catch (e) {
            this.logger.warn(`unable to resolve ${filename}`);
            return [];
        }
    }
}
exports.PushResultFormatter = PushResultFormatter;
const mergeReplacements = (results) => {
    const merged = new Map();
    const replacements = results.filter((result) => result.replacements?.size).map((result) => result.replacements);
    replacements.forEach((replacement) => {
        replacement.forEach((value, key) => {
            if (!merged.has(key)) {
                merged.set(key, value);
            }
            else {
                merged.set(key, Array.from(new Set([...merged.get(key), ...value])));
            }
        });
    });
    return merged;
};
exports.mergeReplacements = mergeReplacements;
