// @salesforce/plugin-community/lib/shared/community/connect/CommunityCreateResource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommunityCreateResource = void 0;
const core_1 = require("@salesforce/core");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-community', 'create');
const NAME_KEY = 'name';
/**
 * A connect api resource for creating a community
 */
class CommunityCreateResource {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    constructor(options) {
        this.options = options;
    }
    // eslint-disable-next-line class-methods-use-this
    handleSuccess(result) {
        const response = {
            message: messages.getMessage('response.createMessage'),
            name: result[NAME_KEY],
            action: messages.getMessage('response.action'),
        };
        return response;
    }
    // eslint-disable-next-line class-methods-use-this
    handleError(error) {
        throw error;
    }
    // eslint-disable-next-line class-methods-use-this
    fetchRelativeConnectUrl() {
        return Promise.resolve('/connect/communities');
    }
    // eslint-disable-next-line class-methods-use-this
    getRequestMethod() {
        return 'POST';
    }
    // eslint-disable-next-line @typescript-eslint/require-await
    async fetchPostParams() {
        const params = {
            name: this.options.name,
            urlPathPrefix: this.options.urlPathPrefix,
            templateName: this.options.templateName,
            description: this.options.description,
            templateParams: this.options['templateParams'],
        };
        return JSON.stringify(params);
    }
}
exports.CommunityCreateResource = CommunityCreateResource;
