// @salesforce/plugin-deploy-retrieve/lib/commands/project/deploy/cancel.js
"use strict";
/*
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@salesforce/core");
const kit_1 = require("@salesforce/kit");
const sf_plugins_core_1 = require("@salesforce/sf-plugins-core");
const source_deploy_retrieve_1 = require("@salesforce/source-deploy-retrieve");
const deploy_1 = require("../../../utils/deploy");
const deployCache_1 = require("../../../utils/deployCache");
const asyncDeployCancelResultFormatter_1 = require("../../../formatters/asyncDeployCancelResultFormatter");
const deployCancelResultFormatter_1 = require("../../../formatters/deployCancelResultFormatter");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-deploy-retrieve', 'deploy.metadata.cancel');
class DeployMetadataCancel extends sf_plugins_core_1.SfCommand {
    async run() {
        const [{ flags }, cache] = await Promise.all([this.parse(DeployMetadataCancel), deployCache_1.DeployCache.create()]);
        const jobId = cache.resolveLatest(flags['use-most-recent'], flags['job-id']);
        // cancel don't care about your tracking conflicts
        const deployOpts = { ...cache.get(jobId), 'ignore-conflicts': true };
        // we may already know the job finished
        if (deployOpts.status &&
            [source_deploy_retrieve_1.RequestStatus.Canceled, source_deploy_retrieve_1.RequestStatus.Failed, source_deploy_retrieve_1.RequestStatus.Succeeded, source_deploy_retrieve_1.RequestStatus.SucceededPartial].includes(deployOpts.status)) {
            messages.createError('error.CannotCancelDeployPre', [jobId, deployOpts.status]);
        }
        if (flags.async) {
            const asyncResult = await (0, deploy_1.cancelDeployAsync)({ 'target-org': deployOpts['target-org'] }, jobId);
            const formatter = new asyncDeployCancelResultFormatter_1.AsyncDeployCancelResultFormatter(asyncResult.id, this.config.bin);
            if (!this.jsonEnabled())
                formatter.display();
            return formatter.getJson();
        }
        else {
            const wait = flags.wait ?? kit_1.Duration.minutes(deployOpts.wait);
            const result = await (0, deploy_1.cancelDeploy)({ ...deployOpts, wait }, jobId);
            const formatter = new deployCancelResultFormatter_1.DeployCancelResultFormatter(result);
            if (!this.jsonEnabled())
                formatter.display();
            cache.update(result.response.id, { status: result.response.status });
            await cache.write();
            if ([source_deploy_retrieve_1.RequestStatus.Succeeded, source_deploy_retrieve_1.RequestStatus.SucceededPartial].includes(result.response.status)) {
                throw messages.createError('error.CannotCancelDeploy');
            }
            return formatter.getJson();
        }
    }
    catch(error) {
        if (error.name.includes('INVALID_ID_FIELD')) {
            const err = messages.createError('error.CannotCancelDeploy');
            return super.catch({ ...error, name: err.name, message: err.message, code: err.code });
        }
        return super.catch(error);
    }
}
exports.default = DeployMetadataCancel;
DeployMetadataCancel.description = messages.getMessage('description');
DeployMetadataCancel.summary = messages.getMessage('summary');
DeployMetadataCancel.examples = messages.getMessages('examples');
DeployMetadataCancel.requiresProject = true;
DeployMetadataCancel.aliases = ['deploy:metadata:cancel'];
DeployMetadataCancel.deprecateAliases = true;
DeployMetadataCancel.flags = {
    async: sf_plugins_core_1.Flags.boolean({
        summary: messages.getMessage('flags.async.summary'),
        description: messages.getMessage('flags.async.description'),
        exclusive: ['wait'],
    }),
    'job-id': sf_plugins_core_1.Flags.salesforceId({
        char: 'i',
        startsWith: '0Af',
        description: messages.getMessage('flags.job-id.description'),
        summary: messages.getMessage('flags.job-id.summary'),
        exactlyOne: ['use-most-recent', 'job-id'],
    }),
    'use-most-recent': sf_plugins_core_1.Flags.boolean({
        char: 'r',
        description: messages.getMessage('flags.use-most-recent.description'),
        summary: messages.getMessage('flags.use-most-recent.summary'),
        exactlyOne: ['use-most-recent', 'job-id'],
    }),
    // we want to allow undefined to use the value from the cache
    // eslint-disable-next-line sf-plugin/flag-min-max-default
    wait: sf_plugins_core_1.Flags.duration({
        char: 'w',
        summary: messages.getMessage('flags.wait.summary'),
        description: messages.getMessage('flags.wait.description'),
        unit: 'minutes',
        helpValue: '<minutes>',
        min: 1,
        exclusive: ['async'],
    }),
};
