// @salesforce/plugin-deploy-retrieve/lib/commands/project/deploy/report.js
"use strict";
/*
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@salesforce/core");
const kit_1 = require("@salesforce/kit");
const sf_plugins_core_1 = require("@salesforce/sf-plugins-core");
const source_deploy_retrieve_1 = require("@salesforce/source-deploy-retrieve");
const deploy_1 = require("../../../utils/deploy");
const deployCache_1 = require("../../../utils/deployCache");
const deployReportResultFormatter_1 = require("../../../formatters/deployReportResultFormatter");
const types_1 = require("../../../utils/types");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-deploy-retrieve', 'deploy.metadata.report');
class DeployMetadataReport extends sf_plugins_core_1.SfCommand {
    async run() {
        const [{ flags }, cache] = await Promise.all([this.parse(DeployMetadataReport), deployCache_1.DeployCache.create()]);
        const jobId = cache.resolveLatest(flags['use-most-recent'], flags['job-id']);
        const deployOpts = cache.get(jobId);
        const org = await core_1.Org.create({ aliasOrUsername: deployOpts['target-org'] });
        const [deployStatus, componentSet] = await Promise.all([
            // we'll use whatever the org supports since we can't specify the org
            // eslint-disable-next-line sf-plugin/get-connection-with-version
            org.getConnection().metadata.checkDeployStatus(jobId, true),
            // if we're using mdapi, we won't have a component set
            deployOpts.isMdapi ? undefined : (0, deploy_1.buildComponentSet)({ ...deployOpts, wait: kit_1.Duration.minutes(deployOpts.wait) }),
        ]);
        const result = new source_deploy_retrieve_1.DeployResult(deployStatus, componentSet);
        const formatter = new deployReportResultFormatter_1.DeployReportResultFormatter(result, {
            ...deployOpts,
            ...{ 'target-org': org },
        });
        if (!this.jsonEnabled())
            formatter.display();
        return formatter.getJson();
    }
}
exports.default = DeployMetadataReport;
DeployMetadataReport.description = messages.getMessage('description');
DeployMetadataReport.summary = messages.getMessage('summary');
DeployMetadataReport.examples = messages.getMessages('examples');
DeployMetadataReport.requiresProject = true;
DeployMetadataReport.aliases = ['deploy:metadata:report'];
DeployMetadataReport.deprecateAliases = true;
DeployMetadataReport.flags = {
    'job-id': sf_plugins_core_1.Flags.salesforceId({
        char: 'i',
        startsWith: '0Af',
        description: messages.getMessage('flags.job-id.description'),
        summary: messages.getMessage('flags.job-id.summary'),
        exactlyOne: ['use-most-recent', 'job-id'],
    }),
    'use-most-recent': sf_plugins_core_1.Flags.boolean({
        char: 'r',
        description: messages.getMessage('flags.use-most-recent.description'),
        summary: messages.getMessage('flags.use-most-recent.summary'),
        exactlyOne: ['use-most-recent', 'job-id'],
    }),
    'coverage-formatters': sf_plugins_core_1.Flags.string({
        multiple: true,
        summary: messages.getMessage('flags.coverage-formatters.summary'),
        options: types_1.reportsFormatters,
    }),
    junit: sf_plugins_core_1.Flags.boolean({ summary: messages.getMessage('flags.junit.summary') }),
    'results-dir': sf_plugins_core_1.Flags.directory({
        dependsOn: ['junit', 'coverage-formatters'],
        summary: messages.getMessage('flags.results-dir.summary'),
    }),
};
