// @salesforce/sf-plugins-core/lib/ux/spinner.js
"use strict";
/*
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Spinner = void 0;
const core_1 = require("@oclif/core");
const _1 = require(".");
/**
 * This class is a light wrapper around ux.action that allows us to
 * automatically suppress any actions if `--json` flag is present.
 */
class Spinner extends _1.UxBase {
    constructor(outputEnabled) {
        super(outputEnabled);
    }
    /**
     * Get the status of the current spinner.
     */
    // eslint-disable-next-line class-methods-use-this
    get status() {
        return core_1.ux.action.status;
    }
    /**
     * Set the status of the current spinner.
     */
    // eslint-disable-next-line class-methods-use-this
    set status(status) {
        core_1.ux.action.status = status;
    }
    /**
     * Start a spinner on the console.
     */
    start(action, status, opts) {
        this.maybeNoop(() => core_1.ux.action.start(action, status, opts));
    }
    /**
     * Stop the spinner on the console.
     */
    stop(msg) {
        this.maybeNoop(() => core_1.ux.action.stop(msg));
    }
    /**
     * Pause the spinner on the console.
     */
    pause(fn, icon) {
        this.maybeNoop(() => core_1.ux.action.pause(fn, icon));
    }
}
exports.Spinner = Spinner;
