// @salesforce/sf-plugins-core/lib/deployer.js
"use strict";
/*
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
/* eslint-disable @typescript-eslint/explicit-member-accessibility */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Deployer = exports.Deployable = void 0;
const events_1 = require("events");
const core_1 = require("@oclif/core");
const ux_1 = require("./ux");
class Deployable {
}
exports.Deployable = Deployable;
/**
 * Interface for deploying Deployables.
 */
class Deployer extends events_1.EventEmitter {
    constructor() {
        super(...arguments);
        /**
         * Deployables are individual pieces that can be deployed on their own. For example,
         * each package in a salesforce project is considered a deployable that can be deployed
         * on its own.
         */
        this.deployables = [];
        this.prompter = new ux_1.Prompter();
    }
    /**
     * Method for displaying deploy progress to the user
     */
    // eslint-disable-next-line @typescript-eslint/no-unused-vars, @typescript-eslint/no-empty-function, class-methods-use-this
    progress(current, total, message) { }
    /**
     * Log messages to the console
     */
    // eslint-disable-next-line class-methods-use-this
    log(msg, ...args) {
        core_1.ux.log(msg, ...args);
    }
    /**
     * Prompt user for additional information
     */
    async prompt(questions, initialAnswers) {
        const answers = await this.prompter.prompt(questions, initialAnswers);
        return answers;
    }
    /**
     * Overwrite the deployables property on the class.
     */
    selectDeployables(deployables) {
        this.deployables = Object.assign([], deployables);
    }
}
exports.Deployer = Deployer;
