// @salesforce/sf-plugins-core/lib/exported.js
"use strict";
/*
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Flags = exports.StandardColors = exports.SfCommand = exports.SfHook = exports.Spinner = exports.Ux = exports.generateTableChoices = exports.Prompter = exports.Progress = exports.Deauthorizer = exports.Deployer = exports.Deployable = exports.parseVarArgs = exports.toHelpSection = void 0;
const core_1 = require("@oclif/core");
var util_1 = require("./util");
Object.defineProperty(exports, "toHelpSection", { enumerable: true, get: function () { return util_1.toHelpSection; } });
Object.defineProperty(exports, "parseVarArgs", { enumerable: true, get: function () { return util_1.parseVarArgs; } });
var deployer_1 = require("./deployer");
Object.defineProperty(exports, "Deployable", { enumerable: true, get: function () { return deployer_1.Deployable; } });
Object.defineProperty(exports, "Deployer", { enumerable: true, get: function () { return deployer_1.Deployer; } });
var deauthorizer_1 = require("./deauthorizer");
Object.defineProperty(exports, "Deauthorizer", { enumerable: true, get: function () { return deauthorizer_1.Deauthorizer; } });
var ux_1 = require("./ux");
Object.defineProperty(exports, "Progress", { enumerable: true, get: function () { return ux_1.Progress; } });
Object.defineProperty(exports, "Prompter", { enumerable: true, get: function () { return ux_1.Prompter; } });
Object.defineProperty(exports, "generateTableChoices", { enumerable: true, get: function () { return ux_1.generateTableChoices; } });
Object.defineProperty(exports, "Ux", { enumerable: true, get: function () { return ux_1.Ux; } });
Object.defineProperty(exports, "Spinner", { enumerable: true, get: function () { return ux_1.Spinner; } });
var hooks_1 = require("./hooks");
Object.defineProperty(exports, "SfHook", { enumerable: true, get: function () { return hooks_1.SfHook; } });
__exportStar(require("./types"), exports);
var sfCommand_1 = require("./sfCommand");
Object.defineProperty(exports, "SfCommand", { enumerable: true, get: function () { return sfCommand_1.SfCommand; } });
Object.defineProperty(exports, "StandardColors", { enumerable: true, get: function () { return sfCommand_1.StandardColors; } });
__exportStar(require("./compatibility"), exports);
__exportStar(require("./stubUx"), exports);
// custom flags
const orgFlags_1 = require("./flags/orgFlags");
const salesforceId_1 = require("./flags/salesforceId");
const orgApiVersion_1 = require("./flags/orgApiVersion");
const duration_1 = require("./flags/duration");
exports.Flags = {
    boolean: core_1.Flags.boolean,
    directory: core_1.Flags.directory,
    file: core_1.Flags.file,
    integer: core_1.Flags.integer,
    string: core_1.Flags.string,
    url: core_1.Flags.url,
    custom: core_1.Flags.custom,
    duration: duration_1.durationFlag,
    salesforceId: salesforceId_1.salesforceIdFlag,
    orgApiVersion: orgApiVersion_1.orgApiVersionFlag,
    requiredOrg: orgFlags_1.requiredOrgFlag,
    requiredHub: orgFlags_1.requiredHubFlag,
    optionalOrg: orgFlags_1.optionalOrgFlag,
    optionalHub: orgFlags_1.optionalHubFlag,
};
