// @salesforce/plugin-packaging/lib/commands/package/install/report.js
"use strict";
/*
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Report = void 0;
const sf_plugins_core_1 = require("@salesforce/sf-plugins-core");
const core_1 = require("@salesforce/core");
const packaging_1 = require("@salesforce/packaging");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-packaging', 'package_install_report');
const installMsgs = core_1.Messages.loadMessages('@salesforce/plugin-packaging', 'package_install');
class Report extends sf_plugins_core_1.SfCommand {
    static parseStatus(binName, request, username, alias) {
        const pkgIdOrAlias = alias ?? request.SubscriberPackageVersionKey;
        const { Status } = request;
        if (Status === 'SUCCESS') {
            return installMsgs.getMessage('package-install-success', [pkgIdOrAlias]);
        }
        else if (['IN_PROGRESS', 'UNKNOWN'].includes(Status)) {
            return installMsgs.getMessage('packageInstallInProgress', [binName, request.Id, username]);
        }
        else {
            let errorMessage = '<empty>';
            const errors = request?.Errors?.errors;
            if (errors?.length) {
                errorMessage = 'Installation errors: ';
                for (let i = 0; i < errors.length; i++) {
                    errorMessage += `\n${i + 1}) ${errors[i].message}`;
                }
            }
            throw installMsgs.createError('packageInstallError', [errorMessage]);
        }
    }
    async run() {
        const { flags } = await this.parse(Report);
        const connection = flags['target-org'].getConnection(flags['api-version']);
        const pkgInstallRequest = await packaging_1.SubscriberPackageVersion.getInstallRequest(flags['request-id'], connection);
        this.log(Report.parseStatus(this.config.bin, pkgInstallRequest, flags['target-org'].getUsername()));
        return pkgInstallRequest;
    }
}
exports.Report = Report;
Report.summary = messages.getMessage('summary');
Report.examples = messages.getMessages('examples');
Report.deprecateAliases = true;
Report.aliases = ['force:package:beta:install:report', 'force:package:install:report'];
Report.flags = {
    'target-org': sf_plugins_core_1.requiredOrgFlagWithDeprecations,
    'api-version': sf_plugins_core_1.orgApiVersionFlagWithDeprecations,
    loglevel: sf_plugins_core_1.loglevel,
    'request-id': sf_plugins_core_1.Flags.salesforceId({
        startsWith: '0Hf',
        length: 'both',
        char: 'i',
        deprecateAliases: true,
        aliases: ['requestid'],
        summary: messages.getMessage('flags.request-id.summary'),
        required: true,
    }),
};
