// @oclif/plugin-command-snapshot/lib/snapshot-command.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SnapshotCommand = void 0;
const core_1 = require("@oclif/core");
const _ = require("lodash");
class SnapshotCommand extends core_1.Command {
    get commands() {
        var _a;
        const devPlugins = (_a = this.config.pjson.oclif.devPlugins) !== null && _a !== void 0 ? _a : [];
        const commands = this.config.commands
            // Ignore dev plugins
            .filter(command => { var _a; return !devPlugins.includes((_a = command.pluginName) !== null && _a !== void 0 ? _a : ''); })
            // remove aliases that reference themselves
            .filter(command => !command.aliases.includes(command.id));
        return _.sortBy(commands, 'id');
    }
    get entries() {
        return this.commands.map(command => {
            return {
                command: command.id,
                plugin: command.pluginName,
                flags: Object.keys(command.flags).sort(),
                alias: command.aliases,
            };
        });
    }
}
exports.SnapshotCommand = SnapshotCommand;
